{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "6c3e7e2a-be98-4a50-aa01-00ae1426fab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv\n",
    "import os\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import scipy.stats as stat\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c2793d34-ea60-48b3-b402-bbfa23df3287",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_maf = pd.read_csv('/Users/lmartin/Documents/2025/Rapid_Autopsy_Convergent_Evolution/Figure2/RA_combined_maf_dedup_updated_on_trees_022025.maf',sep='\\t')\n",
    "subclones_sig_genes = pd.read_csv('/Users/lmartin/Documents/2023/CDK4_mutsig_added_pats/on_subclones_v3/0/mutsig_results/outdir/sig_genes.txt',sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "77258951-8fa7-4eb2-a6ee-9579ea04210c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "subset=(subclones_sig_genes['p'] < 0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "98268e6d-d805-4daa-bd20-84322170959d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6790"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_muts=['Missense_Mutation',\n",
    " 'Nonsense_Mutation',\n",
    " 'Start_Codon_SNP',\n",
    " 'De_novo_Start_InFrame',\n",
    " 'De_novo_Start_OutOfFrame',\n",
    " 'Nonstop_Mutation',\n",
    " 'Frame_Shift_Del',\n",
    " 'Frame_Shift_Ins',\n",
    " 'In_Frame_Del',\n",
    " 'In_Frame_Ins']\n",
    "\n",
    "coding_maf=combined_maf[combined_maf['Variant_Classification'].isin(coding_muts)]\n",
    "len(coding_maf['Hugo_Symbol'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d9143db2-80a7-4c3e-abcd-2833aef722d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_mut_pats</th>\n",
       "      <th>n_mut_clones</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CSAD</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNA2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ESR1</th>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H3B</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IVL</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2C</th>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRAS</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEN1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MNDA</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3CA</th>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ST18</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF536</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             n_mut_pats  n_mut_clones\n",
       "Hugo_Symbol                          \n",
       "CSAD                  2             3\n",
       "CTCF                  2             4\n",
       "CTNNA2                5             5\n",
       "ESR1                  6            10\n",
       "HIST1H3B              2             3\n",
       "IVL                   4             6\n",
       "KMT2C                 5            16\n",
       "KRAS                  3             3\n",
       "MEN1                  1             2\n",
       "MNDA                  3             3\n",
       "PIK3CA               11            15\n",
       "ST18                  4             5\n",
       "TP53                  4             6\n",
       "ZNF536                4             5"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([\n",
    "  coding_maf.groupby([\"Hugo_Symbol\", \"Patient_ID\"]).size().loc[subclones_sig_genes.loc[subset, \"gene\"]].groupby(level = 0).size().rename(\"n_mut_pats\"),\n",
    "  coding_maf.groupby([\"Hugo_Symbol\", \"Tumor_Sample_Barcode\"]).size().loc[subclones_sig_genes.loc[subset, \"gene\"]].groupby(level = 0).size().rename(\"n_mut_clones\")],\n",
    "  axis = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c4aa1b9-54da-4cdc-8f0e-9ebdb1feceaf",
   "metadata": {},
   "source": [
    "## test with KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7dbe4bc6-84e4-4e54-a07c-49b094117396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_Alias</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Cluster_Assignment</th>\n",
       "      <th>Allelic_CN_minor</th>\n",
       "      <th>Allelic_CN_major</th>\n",
       "      <th>preDP_ccf_mean</th>\n",
       "      <th>preDP_ccf_CI_low</th>\n",
       "      <th>preDP_ccf_CI_high</th>\n",
       "      <th>clust_ccf_mean</th>\n",
       "      <th>clust_ccf_CI_low</th>\n",
       "      <th>clust_ccf_CI_high</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>1547</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874772</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>180</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1547_CL15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>1547</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879260</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>184</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1547_CL14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3714</th>\n",
       "      <td>1558</td>\n",
       "      <td>RA_1558_cfDNA_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879322</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>79</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.76</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1558_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>1558</td>\n",
       "      <td>1558_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152012253</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>185</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1558_CL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4071</th>\n",
       "      <td>1326</td>\n",
       "      <td>RP-1700_BWES00001_v1_Exome_OnPrem</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151896451</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>36</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.57</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1326_CL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4208</th>\n",
       "      <td>1326</td>\n",
       "      <td>RP-1700_T01027-3_v4_Exome_OnPrem</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151945382</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>246</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1326_CL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4701</th>\n",
       "      <td>2389</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151891593</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>101</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.90</td>\n",
       "      <td>2389_CL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5037</th>\n",
       "      <td>2389</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151859429</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>185</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5525</th>\n",
       "      <td>2389</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151851222</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>158</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.17</td>\n",
       "      <td>2389_CL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5526</th>\n",
       "      <td>2389</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878626</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>122</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2389_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6603</th>\n",
       "      <td>2389</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878044</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>194</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6704</th>\n",
       "      <td>2389</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874161</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>152</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2389_CL7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6705</th>\n",
       "      <td>2389</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152007146</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6998</th>\n",
       "      <td>2389</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151917668</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>189</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7068</th>\n",
       "      <td>2389</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874215</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7685</th>\n",
       "      <td>2389</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151880082</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>88</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10415</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151871322</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>133</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.64</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>2542_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10416</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878785</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>185</td>\n",
       "      <td>42</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.75</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>2542_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11613</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879670</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.48</td>\n",
       "      <td>2542_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11614</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151932951</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>357</td>\n",
       "      <td>53</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.63</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.98</td>\n",
       "      <td>2542_CL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12389</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152007108</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>234</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.23</td>\n",
       "      <td>2542_CL15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Patient_ID                          Sample_ID  Sample_Alias  \\\n",
       "731          1547                                  6           NaN   \n",
       "854          1547                                 10           NaN   \n",
       "3714         1558                    RA_1558_cfDNA_3           NaN   \n",
       "3957         1558                             1558_6           NaN   \n",
       "4071         1326  RP-1700_BWES00001_v1_Exome_OnPrem           NaN   \n",
       "4208         1326   RP-1700_T01027-3_v4_Exome_OnPrem           NaN   \n",
       "4701         2389                                  1           NaN   \n",
       "5037         2389                                  3           NaN   \n",
       "5525         2389                                  4           NaN   \n",
       "5526         2389                                  4           NaN   \n",
       "6603         2389                                  7           NaN   \n",
       "6704         2389                                  8           NaN   \n",
       "6705         2389                                  8           NaN   \n",
       "6998         2389                                 10           NaN   \n",
       "7068         2389                                 11           NaN   \n",
       "7685         2389                                 18           NaN   \n",
       "10415        2542                             2542_1           NaN   \n",
       "10416        2542                             2542_1           NaN   \n",
       "11613        2542                             2542_2           NaN   \n",
       "11614        2542                             2542_2           NaN   \n",
       "12389        2542                             2542_8           NaN   \n",
       "\n",
       "      Hugo_Symbol  Chromosome  Start_position Reference_Allele  \\\n",
       "731         KMT2C           7       151874772                C   \n",
       "854         KMT2C           7       151879260                C   \n",
       "3714        KMT2C           7       151879322                G   \n",
       "3957        KMT2C           7       152012253                G   \n",
       "4071        KMT2C           7       151896451                T   \n",
       "4208        KMT2C           7       151945382                C   \n",
       "4701        KMT2C           7       151891593                G   \n",
       "5037        KMT2C           7       151859429                C   \n",
       "5525        KMT2C           7       151851222                G   \n",
       "5526        KMT2C           7       151878626                C   \n",
       "6603        KMT2C           7       151878044                G   \n",
       "6704        KMT2C           7       151874161                C   \n",
       "6705        KMT2C           7       152007146                G   \n",
       "6998        KMT2C           7       151917668                G   \n",
       "7068        KMT2C           7       151874215                C   \n",
       "7685        KMT2C           7       151880082                C   \n",
       "10415       KMT2C           7       151871322                C   \n",
       "10416       KMT2C           7       151878785                G   \n",
       "11613       KMT2C           7       151879670                G   \n",
       "11614       KMT2C           7       151932951                C   \n",
       "12389       KMT2C           7       152007108                C   \n",
       "\n",
       "      Tumor_Seq_Allele  t_ref_count  t_alt_count  ... Cluster_Assignment  \\\n",
       "731                  T          180            6  ...                 15   \n",
       "854                  A          184           12  ...                 14   \n",
       "3714                 A           79           33  ...                  1   \n",
       "3957                 C          185           21  ...                 10   \n",
       "4071                 A           36           13  ...                  2   \n",
       "4208                 -          246           64  ...                  9   \n",
       "4701                 A          101           46  ...                  5   \n",
       "5037                 A          185            4  ...                 18   \n",
       "5525                 C          158           10  ...                  4   \n",
       "5526                 A          122            5  ...                 12   \n",
       "6603                 A          194            1  ...                 18   \n",
       "6704                 A          152            1  ...                  7   \n",
       "6705                 A          211            1  ...                 17   \n",
       "6998                 C          189            1  ...                 18   \n",
       "7068                 A           61            1  ...                 18   \n",
       "7685                 A           88            7  ...                 18   \n",
       "10415                A          133           25  ...                 12   \n",
       "10416                A          185           42  ...                 12   \n",
       "11613                A           30            3  ...                 18   \n",
       "11614                T          357           53  ...                  2   \n",
       "12389                G          234            6  ...                 15   \n",
       "\n",
       "      Allelic_CN_minor Allelic_CN_major  preDP_ccf_mean  preDP_ccf_CI_low  \\\n",
       "731                NaN              NaN            0.09              0.04   \n",
       "854                NaN              NaN            0.18              0.10   \n",
       "3714               NaN              NaN            0.89              0.76   \n",
       "3957               NaN              NaN            0.39              0.25   \n",
       "4071               NaN              NaN            0.80              0.57   \n",
       "4208               NaN              NaN            0.87              0.74   \n",
       "4701               NaN              NaN            0.77              0.62   \n",
       "5037               NaN              NaN            0.08              0.03   \n",
       "5525               NaN              NaN            0.22              0.11   \n",
       "5526               NaN              NaN            0.16              0.06   \n",
       "6603               NaN              NaN            0.05              0.01   \n",
       "6704               NaN              NaN            0.38              0.01   \n",
       "6705               NaN              NaN            0.34              0.01   \n",
       "6998               NaN              NaN            0.03              0.01   \n",
       "7068               NaN              NaN            0.06              0.01   \n",
       "7685               NaN              NaN            0.47              0.19   \n",
       "10415              NaN              NaN            0.83              0.64   \n",
       "10416              NaN              NaN            0.88              0.75   \n",
       "11613              NaN              NaN            0.56              0.24   \n",
       "11614              NaN              NaN            0.81              0.63   \n",
       "12389              NaN              NaN            0.39              0.13   \n",
       "\n",
       "       preDP_ccf_CI_high  clust_ccf_mean  clust_ccf_CI_low  clust_ccf_CI_high  \\\n",
       "731                 0.14            0.05              0.05               0.06   \n",
       "854                 0.27            0.06              0.06               0.06   \n",
       "3714                1.00            1.00              0.99               1.00   \n",
       "3957                0.53            0.30              0.29               0.30   \n",
       "4071                1.00            0.98              0.96               1.00   \n",
       "4208                1.00            0.83              0.77               0.89   \n",
       "4701                0.92            0.88              0.86               0.90   \n",
       "5037                0.12            0.01              0.01               0.01   \n",
       "5525                0.36            0.16              0.15               0.17   \n",
       "5526                0.25            0.36              0.35               0.37   \n",
       "6603                0.09            0.01              0.01               0.01   \n",
       "6704                0.82            0.03              0.02               0.03   \n",
       "6705                0.78            0.01              0.01               0.01   \n",
       "6998                0.05            0.01              0.01               0.01   \n",
       "7068                0.10            0.01              0.01               0.01   \n",
       "7685                0.74            0.03              0.03               0.03   \n",
       "10415               1.00            0.86              0.86               0.86   \n",
       "10416               1.00            0.86              0.86               0.86   \n",
       "11613               0.95            0.48              0.47               0.48   \n",
       "11614               1.00            0.94              0.91               0.98   \n",
       "12389               0.64            0.23              0.22               0.23   \n",
       "\n",
       "       Tumor_Sample_Barcode  \n",
       "731               1547_CL15  \n",
       "854               1547_CL14  \n",
       "3714               1558_CL1  \n",
       "3957              1558_CL10  \n",
       "4071               1326_CL2  \n",
       "4208               1326_CL9  \n",
       "4701               2389_CL5  \n",
       "5037              2389_CL18  \n",
       "5525               2389_CL4  \n",
       "5526              2389_CL12  \n",
       "6603              2389_CL18  \n",
       "6704               2389_CL7  \n",
       "6705              2389_CL17  \n",
       "6998              2389_CL18  \n",
       "7068              2389_CL18  \n",
       "7685              2389_CL18  \n",
       "10415             2542_CL12  \n",
       "10416             2542_CL12  \n",
       "11613             2542_CL18  \n",
       "11614              2542_CL2  \n",
       "12389             2542_CL15  \n",
       "\n",
       "[21 rows x 23 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KMT2C_all_muts = coding_maf[coding_maf['Hugo_Symbol']=='KMT2C']\n",
    "KMT2C_all_muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "40eb2b3d-602a-4ee8-ab49-abb4f666ac7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## total number of mutations N0\n",
    "N0 = len(KMT2C_all_muts)\n",
    "## total number of mutated clones N_c\n",
    "N_c = len( KMT2C_all_muts['Tumor_Sample_Barcode'].unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "bb207faa-5e08-4ec0-8735-1db2f8c678b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "f0f17c32-ceed-4478-aa27-ecdb67d9d990",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N_c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc1fb54-96e6-4f61-90ec-965a4821412c",
   "metadata": {},
   "source": [
    "## start with one patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "52bed5fa-6e1c-4bf7-abc3-4aac4962d2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "RA_2389 = coding_maf[coding_maf['Patient_ID']==2389]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af3465be-5842-4af8-b541-bd1ceb4d3cd5",
   "metadata": {},
   "source": [
    "### for every clone, I am going to model the probability that KMT2C is in the clone or not\n",
    "### then I'll tell it what the real distribution is and the sum is the score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "dea32333-95a7-40ef-9f51-da2d0668dce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## total number of mutations N0\n",
    "N0 = len(KMT2C_all_muts)\n",
    "## total number of mutated clones N_c\n",
    "N_c = len( KMT2C_all_muts['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "RA_2389_clones = list(RA_2389['Tumor_Sample_Barcode'].unique() )\n",
    "# total number of mutations in this patient\n",
    "M = len(RA_2389)\n",
    " \n",
    "# initialize a Pij matrix where you have a 1 if a clone was mutated and then the score\n",
    "P_2389 =np.zeros( (len(RA_2389_clones), 2 ))\n",
    "\n",
    "gene=\"KMT2C\"\n",
    "\n",
    "clone_idx=0\n",
    "for clone in RA_2389_clones:\n",
    "    clone_df = RA_2389[RA_2389['Tumor_Sample_Barcode'] == clone]\n",
    "    mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "    Mij = len(list(mutations_in_clone))\n",
    "    \n",
    "    term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "    \n",
    "    # a given clone has no mutation in the gene\n",
    "    P_2389[clone_idx, 0] = 1- term_to_calc\n",
    "    \n",
    "    # a given clone has a mutation in the gene\n",
    "    P_2389[clone_idx, 1] = term_to_calc\n",
    "    \n",
    "\n",
    "    clone_idx +=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "9241ecb8-4e10-4904-b77c-7e70602fe795",
   "metadata": {},
   "outputs": [],
   "source": [
    "## now every row sums to one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "1e39d839-f536-4791-a323-6862ffc0305e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>no_KMT2C</th>\n",
       "      <th>KMT2C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2389_CL16</td>\n",
       "      <td>0.780315</td>\n",
       "      <td>0.219685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2389_CL5</td>\n",
       "      <td>0.471755</td>\n",
       "      <td>0.528245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2389_CL3</td>\n",
       "      <td>0.534833</td>\n",
       "      <td>0.465167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2389_CL18</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2389_CL14</td>\n",
       "      <td>0.584821</td>\n",
       "      <td>0.415179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2389_CL15</td>\n",
       "      <td>0.946495</td>\n",
       "      <td>0.053505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2389_CL1</td>\n",
       "      <td>0.186153</td>\n",
       "      <td>0.813847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2389_CL17</td>\n",
       "      <td>0.846040</td>\n",
       "      <td>0.153960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2389_CL6</td>\n",
       "      <td>0.566707</td>\n",
       "      <td>0.433293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2389_CL7</td>\n",
       "      <td>0.464263</td>\n",
       "      <td>0.535737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2389_CL4</td>\n",
       "      <td>0.281472</td>\n",
       "      <td>0.718528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2389_CL9</td>\n",
       "      <td>0.738166</td>\n",
       "      <td>0.261834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2389_CL13</td>\n",
       "      <td>0.945648</td>\n",
       "      <td>0.054352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2389_CL12</td>\n",
       "      <td>0.688383</td>\n",
       "      <td>0.311617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2389_CL10</td>\n",
       "      <td>0.261257</td>\n",
       "      <td>0.738743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2389_CL11</td>\n",
       "      <td>0.596494</td>\n",
       "      <td>0.403506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        clone  no_KMT2C     KMT2C\n",
       "0   2389_CL16  0.780315  0.219685\n",
       "1    2389_CL5  0.471755  0.528245\n",
       "2    2389_CL3  0.534833  0.465167\n",
       "3   2389_CL18  0.999999  0.000001\n",
       "4   2389_CL14  0.584821  0.415179\n",
       "5   2389_CL15  0.946495  0.053505\n",
       "6    2389_CL1  0.186153  0.813847\n",
       "7   2389_CL17  0.846040  0.153960\n",
       "8    2389_CL6  0.566707  0.433293\n",
       "9    2389_CL7  0.464263  0.535737\n",
       "10   2389_CL4  0.281472  0.718528\n",
       "11   2389_CL9  0.738166  0.261834\n",
       "12  2389_CL13  0.945648  0.054352\n",
       "13  2389_CL12  0.688383  0.311617\n",
       "14  2389_CL10  0.261257  0.738743\n",
       "15  2389_CL11  0.596494  0.403506"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_2389_prob_df = pd.DataFrame(P_2389, columns=['no_KMT2C', 'KMT2C'])\n",
    "P_2389_prob_df.insert(0, 'clone', RA_2389_clones)\n",
    "P_2389_prob_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4caa1ec-5c59-449c-b283-a345e20a6959",
   "metadata": {},
   "source": [
    "## patient score if no KMT2C mutation in any clone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "6af28654-348a-4d47-b470-13b92097a7a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9.259727602330683\n"
     ]
    }
   ],
   "source": [
    "P_2389_prob_df_no_KMT2C = P_2389_prob_df['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C = [ np.log(p) for p in P_2389_prob_df_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8fab121-cf22-4aa9-8d20-1ff5a407d6af",
   "metadata": {},
   "source": [
    "## patient score if a KMT2C mutation in all clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "e3252520-4456-4413-9daa-4eb7ed82fae5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30.67738998828152\n"
     ]
    }
   ],
   "source": [
    "P_2389_prob_df_KMT2C = P_2389_prob_df['KMT2C'].to_list()\n",
    "logs_KMT2C = [ np.log(p) for p in P_2389_prob_df_KMT2C]\n",
    "print( -np.sum(logs_KMT2C))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32568f76-ad33-4f8a-aecf-758a986c41e3",
   "metadata": {},
   "source": [
    "## test RA 1598, no KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "ff0149b2-f0c5-4ab7-adb7-1e4c9f52f7e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "RA_1598= coding_maf[coding_maf['Patient_ID']==1598]\n",
    "RA_1598_clones = list(RA_1598['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "## total number of mutations N0\n",
    "N0 = len(KMT2C_all_muts)\n",
    "## total number of mutated clones N_c\n",
    "N_c = len( KMT2C_all_muts['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "# total number of mutations in this patient\n",
    "M = len(RA_1598)\n",
    "# initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "P_1598 =np.zeros( (len(RA_1598_clones), 2 ))\n",
    "\n",
    "gene=\"KMT2C\"\n",
    "\n",
    "clone_idx=0\n",
    "for clone in RA_1598_clones:\n",
    "    clone_df = RA_1598[RA_1598['Tumor_Sample_Barcode'] == clone]\n",
    "    mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "    Mij = len(list(mutations_in_clone))\n",
    "    term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "    # a given clone has no mutation in the gene\n",
    "    P_1598[clone_idx, 0] = 1- term_to_calc\n",
    "    \n",
    "    #  a given clone has a mutation in the gene\n",
    "    P_1598[clone_idx, 1] = term_to_calc\n",
    "    \n",
    "\n",
    "    clone_idx +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "b8233f8a-9598-4d47-a4a0-6f3d4c550d79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>no_KMT2C</th>\n",
       "      <th>KMT2C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1598_CL2</td>\n",
       "      <td>0.510435</td>\n",
       "      <td>4.895646e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1598_CL1</td>\n",
       "      <td>0.766279</td>\n",
       "      <td>2.337208e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1598_CL8</td>\n",
       "      <td>0.978552</td>\n",
       "      <td>2.144781e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1598_CL10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.977294e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1598_CL5</td>\n",
       "      <td>0.718135</td>\n",
       "      <td>2.818654e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1598_CL4</td>\n",
       "      <td>0.510435</td>\n",
       "      <td>4.895646e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1598_CL7</td>\n",
       "      <td>0.910726</td>\n",
       "      <td>8.927372e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1598_CL9</td>\n",
       "      <td>0.766279</td>\n",
       "      <td>2.337208e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1598_CL3</td>\n",
       "      <td>0.660635</td>\n",
       "      <td>3.393649e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1598_CL6</td>\n",
       "      <td>0.910726</td>\n",
       "      <td>8.927372e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       clone  no_KMT2C         KMT2C\n",
       "0   1598_CL2  0.510435  4.895646e-01\n",
       "1   1598_CL1  0.766279  2.337208e-01\n",
       "2   1598_CL8  0.978552  2.144781e-02\n",
       "3  1598_CL10  1.000000  3.977294e-09\n",
       "4   1598_CL5  0.718135  2.818654e-01\n",
       "5   1598_CL4  0.510435  4.895646e-01\n",
       "6   1598_CL7  0.910726  8.927372e-02\n",
       "7   1598_CL9  0.766279  2.337208e-01\n",
       "8   1598_CL3  0.660635  3.393649e-01\n",
       "9   1598_CL6  0.910726  8.927372e-02"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_1598_prob_df = pd.DataFrame(P_1598, columns=['no_KMT2C', 'KMT2C'])\n",
    "P_1598_prob_df.insert(0, 'clone', RA_1598_clones)\n",
    "P_1598_prob_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16242eda-896b-4dc8-aee1-1f438e56f3a8",
   "metadata": {},
   "source": [
    "## patient score if no KMT2C clones (truth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "abb3dc3b-4b06-4265-9d51-a92e05ee103d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.831758513667644\n"
     ]
    }
   ],
   "source": [
    "P_1598_prob_df_no_KMT2C = P_1598_prob_df['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C_1598 = [ np.log(p) for p in P_1598_prob_df_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C_1598))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ccf5843-6c34-4980-a4c8-cfb48e4a1b0c",
   "metadata": {},
   "source": [
    "## patient score if all KMT2C clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "3465bf8e-6f37-4fb6-9d6f-fa7bdd57feef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34.69963201277903\n"
     ]
    }
   ],
   "source": [
    "P_1598_prob_df_KMT2C = P_1598_prob_df['KMT2C'].to_list()\n",
    "logs_KMT2C_1598 = [ np.log(p) for p in P_1598_prob_df_KMT2C]\n",
    "print( -np.sum(logs_KMT2C_1598))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7e110c6-a2d7-4e39-97ee-2aa317bd4c88",
   "metadata": {},
   "source": [
    "## try for all patients for KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "25f06c03-9d72-40cc-815c-b4816af8e0a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_patients = list(combined_maf['Patient_ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "4c0f1cb5-7e75-4c54-88f3-17209c400de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_probs_KMT2C = []\n",
    "\n",
    "for patient in all_patients: \n",
    "\n",
    "    RA_maf = coding_maf[coding_maf['Patient_ID']==patient]\n",
    "    RA_maf_clones = list(RA_maf['Tumor_Sample_Barcode'].unique() )\n",
    "    \n",
    "    ## total number of mutations N0\n",
    "    N0 = len(KMT2C_all_muts)\n",
    "    ## total number of mutated clones N_c\n",
    "    N_c = len( KMT2C_all_muts['Tumor_Sample_Barcode'].unique() )\n",
    "    \n",
    "    # total number of mutations in this patient\n",
    "    M = len(RA_maf)\n",
    "    # initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "    P_pat =np.zeros( (len(RA_maf_clones), 2 ))\n",
    "    \n",
    "    gene=\"KMT2C\"\n",
    "    \n",
    "    clone_idx=0\n",
    "    for clone in RA_maf_clones:\n",
    "        clone_df = RA_maf[RA_maf['Tumor_Sample_Barcode'] == clone]\n",
    "        mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "        Mij = len(list(mutations_in_clone))\n",
    "        \n",
    "        term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "        # probability that a given clone has a mutation in the gene\n",
    "        P_pat[clone_idx, 1] = term_to_calc\n",
    "        \n",
    "        # probability that a given clone has no mutation in the gene\n",
    "        P_pat[clone_idx, 0] = 1- term_to_calc\n",
    "        clone_idx +=1\n",
    "\n",
    "    P_prob_df = pd.DataFrame(P_pat, columns=['no_KMT2C', 'KMT2C'])\n",
    "    P_prob_df.insert(0, 'clone', RA_maf_clones)\n",
    "    all_probs_KMT2C.append(P_prob_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "a4682501-36e8-4aab-9f2d-94bd54121813",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs_1326 = all_probs[11]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8ac75a0-00cf-464a-befe-8f6dbf6d9022",
   "metadata": {},
   "source": [
    "## score if no KMT2C mutation in this patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "6b9ae65e-4bad-4a09-81c3-f66e005cd019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.969068733986054\n"
     ]
    }
   ],
   "source": [
    "probs_1326_no_KMT2C = probs_1326['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C_1326 = [ np.log(p) for p in probs_1326_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C_1326))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a69120-e34b-472d-bc14-4a6bcbb142d9",
   "metadata": {},
   "source": [
    "## real distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "id": "4ce52f65-5230-4dd1-8915-f1fb2020a9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "KMT2C_clones_RA1326= ['1326_CL2','1326_CL9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "39236529-1bb1-481e-b831-d50c9231b90c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13.4568233637228"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_1326_prob_df_KMT2C_clones = probs_1326[probs_1326['clone'].isin(KMT2C_clones_RA1326)]\n",
    "P_1326_prob_df_non_KMT2C_clones = probs_1326[~probs_1326['clone'].isin(KMT2C_clones_RA1326)]\n",
    "\n",
    "P_1326_prob_df_KMT2C_clones_list = P_1326_prob_df_KMT2C_clones['KMT2C'].to_list()\n",
    "logs_real_KMT2C = [ -np.log(p) for p in P_1326_prob_df_KMT2C_clones_list]\n",
    "\n",
    "P_1326_no_KMT2C_clones_list = P_1326_prob_df_non_KMT2C_clones['no_KMT2C'].to_list()\n",
    "logs_real_no_KMT2C = [ -np.log(p) for p in P_1326_no_KMT2C_clones_list]\n",
    "\n",
    "\n",
    "np.sum( np.concatenate((logs_real_KMT2C, logs_real_no_KMT2C)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "id": "b374f593-d745-4049-9114-f673aba9215d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[        clone  no_KMT2C         KMT2C\n",
       " 0    1547_CL1  0.999841  1.589180e-04\n",
       " 1    1547_CL8  0.660761  3.392387e-01\n",
       " 2    1547_CL2  0.680027  3.199726e-01\n",
       " 3   1547_CL14  1.000000  3.715205e-08\n",
       " 4   1547_CL15  0.974996  2.500380e-02\n",
       " 5   1547_CL12  0.572068  4.279320e-01\n",
       " 6    1547_CL4  0.054017  9.459826e-01\n",
       " 7    1547_CL6  0.105249  8.947513e-01\n",
       " 8    1547_CL3  0.105249  8.947513e-01\n",
       " 9   1547_CL11  0.775481  2.245195e-01\n",
       " 10  1547_CL13  0.054017  9.459826e-01,\n",
       "        clone  no_KMT2C         KMT2C\n",
       " 0   1598_CL2  0.510435  4.895646e-01\n",
       " 1   1598_CL1  0.766279  2.337208e-01\n",
       " 2   1598_CL8  0.978552  2.144781e-02\n",
       " 3  1598_CL10  1.000000  3.977294e-09\n",
       " 4   1598_CL5  0.718135  2.818654e-01\n",
       " 5   1598_CL4  0.510435  4.895646e-01\n",
       " 6   1598_CL7  0.910726  8.927372e-02\n",
       " 7   1598_CL9  0.766279  2.337208e-01\n",
       " 8   1598_CL3  0.660635  3.393649e-01\n",
       " 9   1598_CL6  0.910726  8.927372e-02,\n",
       "       clone  no_KMT2C         KMT2C\n",
       " 0  1078_CL1  0.999918  8.249165e-05\n",
       " 1  1078_CL5  0.997347  2.652554e-03\n",
       " 2  1078_CL2  0.834276  1.657242e-01\n",
       " 3  1078_CL6  1.000000  3.030528e-09\n",
       " 4  1078_CL4  0.293469  7.065306e-01,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0    1045_CL9  0.747448  0.252552\n",
       " 1    1045_CL1  0.818579  0.181421\n",
       " 2    1045_CL3  0.818579  0.181421\n",
       " 3   1045_CL16  0.999997  0.000003\n",
       " 4    1045_CL4  0.404686  0.595314\n",
       " 5   1045_CL10  0.650226  0.349774\n",
       " 6   1045_CL14  0.464179  0.535821\n",
       " 7   1045_CL13  0.995193  0.004807\n",
       " 8   1045_CL15  0.963975  0.036025\n",
       " 9    1045_CL7  0.338935  0.661065\n",
       " 10  1045_CL11  0.907873  0.092127\n",
       " 11   1045_CL6  0.097615  0.902385\n",
       " 12  1045_CL12  0.338935  0.661065,\n",
       "        clone  no_KMT2C     KMT2C\n",
       " 0  1113_CL13  0.448316  0.551684\n",
       " 1   1113_CL1  0.999316  0.000684\n",
       " 2   1113_CL8  0.940401  0.059599\n",
       " 3   1113_CL3  0.840768  0.159232\n",
       " 4   1113_CL2  0.940401  0.059599\n",
       " 5   1113_CL5  0.700876  0.299124\n",
       " 6  1113_CL12  0.992812  0.007188\n",
       " 7  1113_CL11  0.999550  0.000450\n",
       " 8  1113_CL10  0.700876  0.299124,\n",
       "       clone  no_KMT2C         KMT2C\n",
       " 0  1644_CL3  0.999787  2.127926e-04\n",
       " 1  1644_CL1  1.000000  2.605549e-09\n",
       " 2  1644_CL8  0.905056  9.494445e-02\n",
       " 3  1644_CL6  0.992019  7.981029e-03\n",
       " 4  1644_CL7  0.431261  5.687392e-01,\n",
       "        clone  no_KMT2C         KMT2C\n",
       " 0   2819_CL6  0.390469  6.095313e-01\n",
       " 1   2819_CL1  0.893010  1.069901e-01\n",
       " 2  2819_CL10  1.000000  7.006126e-08\n",
       " 3   2819_CL4  0.280172  7.198278e-01\n",
       " 4  2819_CL11  0.925801  7.419911e-02\n",
       " 5   2819_CL8  0.816901  1.830986e-01\n",
       " 6   2819_CL5  0.957669  4.233148e-02\n",
       " 7   2819_CL9  0.996191  3.809284e-03\n",
       " 8   2819_CL7  0.390469  6.095313e-01\n",
       " 9  2819_CL12  0.218135  7.818646e-01,\n",
       "       clone  no_KMT2C     KMT2C\n",
       " 0  2974_CL2  0.921519  0.078481\n",
       " 1  2974_CL8  0.999992  0.000008\n",
       " 2  2974_CL1  0.994742  0.005258\n",
       " 3  2974_CL4  0.600914  0.399086\n",
       " 4  2974_CL9  0.999633  0.000367\n",
       " 5  2974_CL6  0.365087  0.634913\n",
       " 6  2974_CL3  0.365087  0.634913\n",
       " 7  2974_CL5  0.847285  0.152715,\n",
       "        clone  no_KMT2C     KMT2C\n",
       " 0   1035_CL1  0.999736  0.000264\n",
       " 1   1035_CL5  0.942306  0.057694\n",
       " 2   1035_CL6  0.959028  0.040972\n",
       " 3   1035_CL3  0.451884  0.548116\n",
       " 4   1035_CL7  0.887451  0.112549\n",
       " 5   1035_CL4  0.258056  0.741944\n",
       " 6   1035_CL2  0.258056  0.741944\n",
       " 7  1035_CL10  0.999055  0.000945\n",
       " 8   1035_CL9  0.990173  0.009827\n",
       " 9   1035_CL8  0.704847  0.295153,\n",
       "       clone  no_KMT2C     KMT2C\n",
       " 0  1534_CL1  1.000000  0.000000\n",
       " 1  1534_CL4  0.978263  0.021737\n",
       " 2  1534_CL2  0.900741  0.099259\n",
       " 3  1534_CL3  0.590882  0.409118,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0    1558_CL1  0.956235  0.043765\n",
       " 1    1558_CL5  0.277839  0.722161\n",
       " 2    1558_CL6  0.535837  0.464163\n",
       " 3   1558_CL11  0.999996  0.000004\n",
       " 4    1558_CL4  0.420361  0.579639\n",
       " 5    1558_CL3  0.481150  0.518850\n",
       " 6    1558_CL2  0.420361  0.579639\n",
       " 7    1558_CL9  0.790736  0.209264\n",
       " 8    1558_CL8  0.629189  0.370811\n",
       " 9   1558_CL10  0.999991  0.000009\n",
       " 10   1558_CL7  0.277839  0.722161,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0   1326_CL13  0.999998  0.000002\n",
       " 1    1326_CL6  0.834276  0.165724\n",
       " 2    1326_CL1  0.927014  0.072986\n",
       " 3   1326_CL10  0.635123  0.364877\n",
       " 4    1326_CL2  0.958520  0.041480\n",
       " 5    1326_CL3  0.219287  0.780713\n",
       " 6    1326_CL7  0.528375  0.471625\n",
       " 7    1326_CL9  0.987227  0.012773\n",
       " 8    1326_CL5  0.219287  0.780713\n",
       " 9   1326_CL12  0.987227  0.012773\n",
       " 10   1326_CL8  0.219287  0.780713,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0   2389_CL16  0.780315  0.219685\n",
       " 1    2389_CL5  0.471755  0.528245\n",
       " 2    2389_CL3  0.534833  0.465167\n",
       " 3   2389_CL18  0.999999  0.000001\n",
       " 4   2389_CL14  0.584821  0.415179\n",
       " 5   2389_CL15  0.946495  0.053505\n",
       " 6    2389_CL1  0.186153  0.813847\n",
       " 7   2389_CL17  0.846040  0.153960\n",
       " 8    2389_CL6  0.566707  0.433293\n",
       " 9    2389_CL7  0.464263  0.535737\n",
       " 10   2389_CL4  0.281472  0.718528\n",
       " 11   2389_CL9  0.738166  0.261834\n",
       " 12  2389_CL13  0.945648  0.054352\n",
       " 13  2389_CL12  0.688383  0.311617\n",
       " 14  2389_CL10  0.261257  0.738743\n",
       " 15  2389_CL11  0.596494  0.403506,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0   2542_CL12  0.999999  0.000001\n",
       " 1   2542_CL10  0.089337  0.910663\n",
       " 2    2542_CL1  0.789641  0.210359\n",
       " 3    2542_CL3  0.528486  0.471514\n",
       " 4   2542_CL18  0.911691  0.088309\n",
       " 5   2542_CL16  0.995585  0.004415\n",
       " 6   2542_CL15  0.850428  0.149572\n",
       " 7    2542_CL2  0.192082  0.807918\n",
       " 8   2542_CL17  0.540776  0.459224\n",
       " 9    2542_CL9  0.119897  0.880103\n",
       " 10   2542_CL7  0.097071  0.902929\n",
       " 11  2542_CL13  0.376335  0.623665\n",
       " 12   2542_CL6  0.271269  0.728731\n",
       " 13   2542_CL5  0.008454  0.991546\n",
       " 14  2542_CL14  0.713600  0.286400\n",
       " 15  2542_CL11  0.073681  0.926319\n",
       " 16  2542_CL19  0.718717  0.281283\n",
       " 17   2542_CL4  0.081541  0.918459\n",
       " 18   2542_CL8  0.008454  0.991546,\n",
       "         clone  no_KMT2C         KMT2C\n",
       " 0    1002_CL1  0.322656  6.773436e-01\n",
       " 1   1002_CL21  1.000000  1.578845e-12\n",
       " 2   1002_CL11  0.083166  9.168338e-01\n",
       " 3    1002_CL3  0.118752  8.812484e-01\n",
       " 4    1002_CL8  0.254351  7.456494e-01\n",
       " 5   1002_CL10  0.753544  2.464563e-01\n",
       " 6   1002_CL20  0.951601  4.839921e-02\n",
       " 7   1002_CL19  0.529440  4.705603e-01\n",
       " 8   1002_CL16  0.501800  4.982005e-01\n",
       " 9    1002_CL6  0.046215  9.537852e-01\n",
       " 10  1002_CL18  0.369930  6.300703e-01\n",
       " 11  1002_CL15  0.248372  7.516285e-01\n",
       " 12  1002_CL13  0.254351  7.456494e-01\n",
       " 13  1002_CL12  0.505841  4.941593e-01\n",
       " 14   1002_CL5  0.068552  9.314477e-01\n",
       " 15  1002_CL17  0.666995  3.330047e-01\n",
       " 16  1002_CL14  0.192466  8.075340e-01]"
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "605b99ce-4506-4f86-9d1f-8e9ec58a6ec5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_probs_order = [ 1547, 1598, 1078, 1045, 1113, 1644, 2819, 2974, 1035,1534, 1558, 1326, 2389, 2542, 1002  ]\n",
    "len(all_probs_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "30603c03-8872-4ba5-83f8-b52293d4581c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pats_with_no_KMT2C = [1598, 1078, 1045, 1113, 1644, 2819, 2974, 1035,1534,1002]\n",
    "pats_with_no_KMT2C_idx =[1, 2, 3, 4, 5, 6, 7, 8, 9, 14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "b8956df6-9b6b-412a-b0a0-e71d74e31b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs_pats_no_KMT2C = []\n",
    "probs_pats_no_KMT2C_scores = []\n",
    "probs_pats_no_KMT2C_scores_dict = {}\n",
    "for idx in pats_with_no_KMT2C_idx:\n",
    "    score = 0\n",
    "    probs_pats_no_KMT2C.append(all_probs[idx])\n",
    "    pat_probs = all_probs[idx]\n",
    "    pat_probs_no_KMT2C = pat_probs['no_KMT2C'].to_list()\n",
    "    logs_no_KMT2C = [ np.log(p) for p in pat_probs_no_KMT2C]\n",
    "    score = -np.sum(logs_no_KMT2C)\n",
    "    \n",
    "    probs_pats_no_KMT2C_scores.append(score)\n",
    "    pat = all_probs_order[idx]\n",
    "    probs_pats_no_KMT2C_scores_dict[pat] = score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "f872a4f5-6500-4467-bc1c-7454eaa81afa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1598: 2.831758513667644,\n",
       " 1078: 1.4099116376412282,\n",
       " 1045: 7.422806015861028,\n",
       " 1113: 1.8177880568251152,\n",
       " 1644: 0.9490269803022182,\n",
       " 2819: 5.1153659950758,\n",
       " 2974: 2.7776425563567213,\n",
       " 1035: 4.085005605081089,\n",
       " 1534: 0.6526531549056974,\n",
       " 1002: 21.021183275023688}"
      ]
     },
     "execution_count": 283,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs_pats_no_KMT2C_scores_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ff2b34-0561-4d69-8d2c-a78039c9e155",
   "metadata": {},
   "outputs": [],
   "source": [
    "## maybe apply monotonization for RA 1002?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "a0ed914d-ec21-466d-bd21-e57ed1156d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "pats_with_KMT2C_idx=[0, 10,11,12,13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "708331f5-7dad-4ee2-bbc3-13cad03247ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_KMT2C_clones = list(KMT2C_all_muts['Tumor_Sample_Barcode'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "ab44785a-578a-4620-9321-7b8197f33edc",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs_pats_with_KMT2C = []\n",
    "probs_pats_KMT2C_scores = []\n",
    "probs_pats_with_KMT2C_scores_dict = {}\n",
    "for idx in pats_with_KMT2C_idx:\n",
    "    #score=0\n",
    "    running_score=[]\n",
    "    pat_probs = all_probs[idx]\n",
    "    all_clones_in_pat = list(pat_probs['clone'])\n",
    "    for clone in all_clones_in_pat:\n",
    "        pat_probs_clone = pat_probs[pat_probs['clone']==clone]\n",
    "        if clone in all_KMT2C_clones:\n",
    "\n",
    "            pat_probs_clone_KMT2C = pat_probs_clone['KMT2C'].to_list()\n",
    "            logs_KMT2C = [ - np.log(p) for p in pat_probs_clone_KMT2C]\n",
    "            running_score.extend(logs_KMT2C)\n",
    "        else:\n",
    "            pat_probs_no_KMT2C = pat_probs_clone['no_KMT2C'].to_list()\n",
    "            logs_no_KMT2C = [ - np.log(p) for p in pat_probs_no_KMT2C]\n",
    "            running_score.extend(logs_no_KMT2C)\n",
    "            \n",
    "    score = np.sum(running_score)\n",
    "    \n",
    "    probs_pats_with_KMT2C.append(pat_probs)\n",
    "    probs_pats_KMT2C_scores.append(score)\n",
    "    pat = all_probs_order[idx]\n",
    "    probs_pats_with_KMT2C_scores_dict[pat] = score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "458ad730-0f9e-4ba0-bd9c-fcedca07be6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1547: 32.749645292558064,\n",
       " 1558: 21.11796516113077,\n",
       " 1326: 13.456823363722798,\n",
       " 2389: 24.01724717124339,\n",
       " 2542: 44.20458330060845}"
      ]
     },
     "execution_count": 289,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs_pats_with_KMT2C_scores_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "e7893ffc-d06e-4963-a0cb-6bcbcdc22947",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[32.749645292558064,\n",
       " 21.11796516113077,\n",
       " 13.456823363722798,\n",
       " 24.01724717124339,\n",
       " 44.20458330060845]"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs_pats_KMT2C_scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af8a38a-bab1-4852-885b-23145742a166",
   "metadata": {},
   "source": [
    "## test any gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9969e7cb-1f17-49dd-ad95-741ac1b55640",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_distributions(gene):\n",
    "\n",
    "    gene_all_muts = coding_maf[coding_maf['Hugo_Symbol']==gene]\n",
    "    all_probs = []\n",
    "    \n",
    "    for patient in all_patients: \n",
    "    \n",
    "        RA_maf = coding_maf[coding_maf['Patient_ID']==patient]\n",
    "        RA_maf_clones = list(RA_maf['Tumor_Sample_Barcode'].unique() )\n",
    "        \n",
    "        ## total number of mutations N0\n",
    "        N0 = len(gene_all_muts)\n",
    "        ## total number of mutated clones N_c\n",
    "        N_c = len( gene_all_muts['Tumor_Sample_Barcode'].unique() )\n",
    "        \n",
    "        # total number of mutations in this patient\n",
    "        M = len(RA_maf)\n",
    "        # initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "        P_pat =np.zeros( (len(RA_maf_clones), 2 ))\n",
    "        \n",
    "        clone_idx=0\n",
    "        for clone in RA_maf_clones:\n",
    "            clone_df = RA_maf[RA_maf['Tumor_Sample_Barcode'] == clone]\n",
    "            mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "            Mij = len(list(mutations_in_clone))\n",
    "            term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "\n",
    "            # probability that a given clone has no mutation in the gene\n",
    "            P_pat[clone_idx, 0] = 1- term_to_calc\n",
    "            \n",
    "            \n",
    "            # probability that a given clone has a mutation in the gene\n",
    "            P_pat[clone_idx, 1] = term_to_calc\n",
    "            clone_idx +=1\n",
    "\n",
    "    \n",
    "        P_prob_df = pd.DataFrame(P_pat, columns=['no_gene_mutations', 'gene_mutation'])\n",
    "        P_prob_df.insert(0, 'clone', RA_maf_clones)\n",
    "        all_probs.append(P_prob_df)\n",
    "    return all_probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "65f550ed-4630-47af-9f0f-900274af9576",
   "metadata": {},
   "outputs": [],
   "source": [
    "esr1_Pm = calculate_distributions('ESR1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "812883a7-f186-4944-a61f-128cafc538bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
