{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6c3e7e2a-be98-4a50-aa01-00ae1426fab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv\n",
    "import os\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import scipy.stats as stat\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2793d34-ea60-48b3-b402-bbfa23df3287",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_maf = pd.read_csv('/Users/lmartin/Documents/2025/RA_combined_maf_dedup_updated_on_trees_022025.maf',sep='\\t')\n",
    "subclones_sig_genes = pd.read_csv('/Users/lmartin/Documents/2023/CDK4_mutsig_added_pats/on_subclones_v3/0/mutsig_results/outdir/sig_genes.txt',sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "77258951-8fa7-4eb2-a6ee-9579ea04210c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "subset=(subclones_sig_genes['p'] < 0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98268e6d-d805-4daa-bd20-84322170959d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6790"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_muts=['Missense_Mutation',\n",
    " 'Nonsense_Mutation',\n",
    " 'Start_Codon_SNP',\n",
    " 'De_novo_Start_InFrame',\n",
    " 'De_novo_Start_OutOfFrame',\n",
    " 'Nonstop_Mutation',\n",
    " 'Frame_Shift_Del',\n",
    " 'Frame_Shift_Ins',\n",
    " 'In_Frame_Del',\n",
    " 'In_Frame_Ins']\n",
    "\n",
    "coding_maf=combined_maf[combined_maf['Variant_Classification'].isin(coding_muts)]\n",
    "len(coding_maf['Hugo_Symbol'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d9143db2-80a7-4c3e-abcd-2833aef722d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_mut_pats</th>\n",
       "      <th>n_mut_clones</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CSAD</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNA2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ESR1</th>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H3B</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IVL</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2C</th>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRAS</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEN1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MNDA</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3CA</th>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ST18</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF536</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             n_mut_pats  n_mut_clones\n",
       "Hugo_Symbol                          \n",
       "CSAD                  2             3\n",
       "CTCF                  2             4\n",
       "CTNNA2                5             5\n",
       "ESR1                  6            10\n",
       "HIST1H3B              2             3\n",
       "IVL                   4             6\n",
       "KMT2C                 5            16\n",
       "KRAS                  3             3\n",
       "MEN1                  1             2\n",
       "MNDA                  3             3\n",
       "PIK3CA               11            15\n",
       "ST18                  4             5\n",
       "TP53                  4             6\n",
       "ZNF536                4             5"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([\n",
    "  coding_maf.groupby([\"Hugo_Symbol\", \"Patient_ID\"]).size().loc[subclones_sig_genes.loc[subset, \"gene\"]].groupby(level = 0).size().rename(\"n_mut_pats\"),\n",
    "  coding_maf.groupby([\"Hugo_Symbol\", \"Tumor_Sample_Barcode\"]).size().loc[subclones_sig_genes.loc[subset, \"gene\"]].groupby(level = 0).size().rename(\"n_mut_clones\")],\n",
    "  axis = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c4aa1b9-54da-4cdc-8f0e-9ebdb1feceaf",
   "metadata": {},
   "source": [
    "## test with KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7dbe4bc6-84e4-4e54-a07c-49b094117396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_Alias</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Cluster_Assignment</th>\n",
       "      <th>Allelic_CN_minor</th>\n",
       "      <th>Allelic_CN_major</th>\n",
       "      <th>preDP_ccf_mean</th>\n",
       "      <th>preDP_ccf_CI_low</th>\n",
       "      <th>preDP_ccf_CI_high</th>\n",
       "      <th>clust_ccf_mean</th>\n",
       "      <th>clust_ccf_CI_low</th>\n",
       "      <th>clust_ccf_CI_high</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>1547</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874772</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>180</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1547_CL15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>1547</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879260</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>184</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1547_CL14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3714</th>\n",
       "      <td>1558</td>\n",
       "      <td>RA_1558_cfDNA_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879322</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>79</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.76</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1558_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>1558</td>\n",
       "      <td>1558_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152012253</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>185</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1558_CL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4071</th>\n",
       "      <td>1326</td>\n",
       "      <td>RP-1700_BWES00001_v1_Exome_OnPrem</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151896451</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>36</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.57</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1326_CL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4208</th>\n",
       "      <td>1326</td>\n",
       "      <td>RP-1700_T01027-3_v4_Exome_OnPrem</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151945382</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>246</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1326_CL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4701</th>\n",
       "      <td>2389</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151891593</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>101</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.90</td>\n",
       "      <td>2389_CL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5037</th>\n",
       "      <td>2389</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151859429</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>185</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5525</th>\n",
       "      <td>2389</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151851222</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>158</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.17</td>\n",
       "      <td>2389_CL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5526</th>\n",
       "      <td>2389</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878626</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>122</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2389_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6603</th>\n",
       "      <td>2389</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878044</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>194</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6704</th>\n",
       "      <td>2389</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874161</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>152</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2389_CL7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6705</th>\n",
       "      <td>2389</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152007146</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6998</th>\n",
       "      <td>2389</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151917668</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>189</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7068</th>\n",
       "      <td>2389</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874215</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7685</th>\n",
       "      <td>2389</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151880082</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>88</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10415</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151871322</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>133</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.64</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>2542_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10416</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878785</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>185</td>\n",
       "      <td>42</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.75</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>2542_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11613</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879670</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.48</td>\n",
       "      <td>2542_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11614</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151932951</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>357</td>\n",
       "      <td>53</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.63</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.98</td>\n",
       "      <td>2542_CL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12389</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152007108</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>234</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.23</td>\n",
       "      <td>2542_CL15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Patient_ID                          Sample_ID  Sample_Alias  \\\n",
       "731          1547                                  6           NaN   \n",
       "854          1547                                 10           NaN   \n",
       "3714         1558                    RA_1558_cfDNA_3           NaN   \n",
       "3957         1558                             1558_6           NaN   \n",
       "4071         1326  RP-1700_BWES00001_v1_Exome_OnPrem           NaN   \n",
       "4208         1326   RP-1700_T01027-3_v4_Exome_OnPrem           NaN   \n",
       "4701         2389                                  1           NaN   \n",
       "5037         2389                                  3           NaN   \n",
       "5525         2389                                  4           NaN   \n",
       "5526         2389                                  4           NaN   \n",
       "6603         2389                                  7           NaN   \n",
       "6704         2389                                  8           NaN   \n",
       "6705         2389                                  8           NaN   \n",
       "6998         2389                                 10           NaN   \n",
       "7068         2389                                 11           NaN   \n",
       "7685         2389                                 18           NaN   \n",
       "10415        2542                             2542_1           NaN   \n",
       "10416        2542                             2542_1           NaN   \n",
       "11613        2542                             2542_2           NaN   \n",
       "11614        2542                             2542_2           NaN   \n",
       "12389        2542                             2542_8           NaN   \n",
       "\n",
       "      Hugo_Symbol  Chromosome  Start_position Reference_Allele  \\\n",
       "731         KMT2C           7       151874772                C   \n",
       "854         KMT2C           7       151879260                C   \n",
       "3714        KMT2C           7       151879322                G   \n",
       "3957        KMT2C           7       152012253                G   \n",
       "4071        KMT2C           7       151896451                T   \n",
       "4208        KMT2C           7       151945382                C   \n",
       "4701        KMT2C           7       151891593                G   \n",
       "5037        KMT2C           7       151859429                C   \n",
       "5525        KMT2C           7       151851222                G   \n",
       "5526        KMT2C           7       151878626                C   \n",
       "6603        KMT2C           7       151878044                G   \n",
       "6704        KMT2C           7       151874161                C   \n",
       "6705        KMT2C           7       152007146                G   \n",
       "6998        KMT2C           7       151917668                G   \n",
       "7068        KMT2C           7       151874215                C   \n",
       "7685        KMT2C           7       151880082                C   \n",
       "10415       KMT2C           7       151871322                C   \n",
       "10416       KMT2C           7       151878785                G   \n",
       "11613       KMT2C           7       151879670                G   \n",
       "11614       KMT2C           7       151932951                C   \n",
       "12389       KMT2C           7       152007108                C   \n",
       "\n",
       "      Tumor_Seq_Allele  t_ref_count  t_alt_count  ... Cluster_Assignment  \\\n",
       "731                  T          180            6  ...                 15   \n",
       "854                  A          184           12  ...                 14   \n",
       "3714                 A           79           33  ...                  1   \n",
       "3957                 C          185           21  ...                 10   \n",
       "4071                 A           36           13  ...                  2   \n",
       "4208                 -          246           64  ...                  9   \n",
       "4701                 A          101           46  ...                  5   \n",
       "5037                 A          185            4  ...                 18   \n",
       "5525                 C          158           10  ...                  4   \n",
       "5526                 A          122            5  ...                 12   \n",
       "6603                 A          194            1  ...                 18   \n",
       "6704                 A          152            1  ...                  7   \n",
       "6705                 A          211            1  ...                 17   \n",
       "6998                 C          189            1  ...                 18   \n",
       "7068                 A           61            1  ...                 18   \n",
       "7685                 A           88            7  ...                 18   \n",
       "10415                A          133           25  ...                 12   \n",
       "10416                A          185           42  ...                 12   \n",
       "11613                A           30            3  ...                 18   \n",
       "11614                T          357           53  ...                  2   \n",
       "12389                G          234            6  ...                 15   \n",
       "\n",
       "      Allelic_CN_minor Allelic_CN_major  preDP_ccf_mean  preDP_ccf_CI_low  \\\n",
       "731                NaN              NaN            0.09              0.04   \n",
       "854                NaN              NaN            0.18              0.10   \n",
       "3714               NaN              NaN            0.89              0.76   \n",
       "3957               NaN              NaN            0.39              0.25   \n",
       "4071               NaN              NaN            0.80              0.57   \n",
       "4208               NaN              NaN            0.87              0.74   \n",
       "4701               NaN              NaN            0.77              0.62   \n",
       "5037               NaN              NaN            0.08              0.03   \n",
       "5525               NaN              NaN            0.22              0.11   \n",
       "5526               NaN              NaN            0.16              0.06   \n",
       "6603               NaN              NaN            0.05              0.01   \n",
       "6704               NaN              NaN            0.38              0.01   \n",
       "6705               NaN              NaN            0.34              0.01   \n",
       "6998               NaN              NaN            0.03              0.01   \n",
       "7068               NaN              NaN            0.06              0.01   \n",
       "7685               NaN              NaN            0.47              0.19   \n",
       "10415              NaN              NaN            0.83              0.64   \n",
       "10416              NaN              NaN            0.88              0.75   \n",
       "11613              NaN              NaN            0.56              0.24   \n",
       "11614              NaN              NaN            0.81              0.63   \n",
       "12389              NaN              NaN            0.39              0.13   \n",
       "\n",
       "       preDP_ccf_CI_high  clust_ccf_mean  clust_ccf_CI_low  clust_ccf_CI_high  \\\n",
       "731                 0.14            0.05              0.05               0.06   \n",
       "854                 0.27            0.06              0.06               0.06   \n",
       "3714                1.00            1.00              0.99               1.00   \n",
       "3957                0.53            0.30              0.29               0.30   \n",
       "4071                1.00            0.98              0.96               1.00   \n",
       "4208                1.00            0.83              0.77               0.89   \n",
       "4701                0.92            0.88              0.86               0.90   \n",
       "5037                0.12            0.01              0.01               0.01   \n",
       "5525                0.36            0.16              0.15               0.17   \n",
       "5526                0.25            0.36              0.35               0.37   \n",
       "6603                0.09            0.01              0.01               0.01   \n",
       "6704                0.82            0.03              0.02               0.03   \n",
       "6705                0.78            0.01              0.01               0.01   \n",
       "6998                0.05            0.01              0.01               0.01   \n",
       "7068                0.10            0.01              0.01               0.01   \n",
       "7685                0.74            0.03              0.03               0.03   \n",
       "10415               1.00            0.86              0.86               0.86   \n",
       "10416               1.00            0.86              0.86               0.86   \n",
       "11613               0.95            0.48              0.47               0.48   \n",
       "11614               1.00            0.94              0.91               0.98   \n",
       "12389               0.64            0.23              0.22               0.23   \n",
       "\n",
       "       Tumor_Sample_Barcode  \n",
       "731               1547_CL15  \n",
       "854               1547_CL14  \n",
       "3714               1558_CL1  \n",
       "3957              1558_CL10  \n",
       "4071               1326_CL2  \n",
       "4208               1326_CL9  \n",
       "4701               2389_CL5  \n",
       "5037              2389_CL18  \n",
       "5525               2389_CL4  \n",
       "5526              2389_CL12  \n",
       "6603              2389_CL18  \n",
       "6704               2389_CL7  \n",
       "6705              2389_CL17  \n",
       "6998              2389_CL18  \n",
       "7068              2389_CL18  \n",
       "7685              2389_CL18  \n",
       "10415             2542_CL12  \n",
       "10416             2542_CL12  \n",
       "11613             2542_CL18  \n",
       "11614              2542_CL2  \n",
       "12389             2542_CL15  \n",
       "\n",
       "[21 rows x 23 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KMT2C_all_muts = coding_maf[coding_maf['Hugo_Symbol']=='KMT2C']\n",
    "KMT2C_all_muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "40eb2b3d-602a-4ee8-ab49-abb4f666ac7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## total number of mutations in the target gene N0\n",
    "N0 = len(KMT2C_all_muts)\n",
    "## total number of clones\n",
    "N_c = len( coding_maf['Tumor_Sample_Barcode'].unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bb207faa-5e08-4ec0-8735-1db2f8c678b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f0f17c32-ceed-4478-aa27-ecdb67d9d990",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "159"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N_c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc1fb54-96e6-4f61-90ec-965a4821412c",
   "metadata": {},
   "source": [
    "## start with one patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "52bed5fa-6e1c-4bf7-abc3-4aac4962d2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "RA_2389 = coding_maf[coding_maf['Patient_ID']==2389]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af3465be-5842-4af8-b541-bd1ceb4d3cd5",
   "metadata": {},
   "source": [
    "### for every clone, I am going to model the probability that KMT2C is in the clone or not\n",
    "### then I'll tell it what the real distribution is and the sum is the score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "dea32333-95a7-40ef-9f51-da2d0668dce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "RA_2389_clones = list(RA_2389['Tumor_Sample_Barcode'].unique() )\n",
    "# total number of mutations in this patient\n",
    "M = len(RA_2389)\n",
    " \n",
    "# initialize a Pij matrix where you have a 1 if a clone was mutated and then the score\n",
    "P_2389 =np.zeros( (len(RA_2389_clones), 2 ))\n",
    "\n",
    "gene=\"KMT2C\"\n",
    "\n",
    "clone_idx=0\n",
    "for clone in RA_2389_clones:\n",
    "    clone_df = RA_2389[RA_2389['Tumor_Sample_Barcode'] == clone]\n",
    "    mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "    Mij = len(list(mutations_in_clone))\n",
    "    \n",
    "    term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "    \n",
    "    # a given clone has no mutation in the gene\n",
    "    P_2389[clone_idx, 0] =  term_to_calc\n",
    "    \n",
    "    # a given clone has one or more mutations in the gene\n",
    "    P_2389[clone_idx, 1] = 1- term_to_calc\n",
    "    \n",
    "\n",
    "    clone_idx +=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9241ecb8-4e10-4904-b77c-7e70602fe795",
   "metadata": {},
   "outputs": [],
   "source": [
    "## now every row sums to one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2ed36da2-9c77-42f9-ab5e-c066a65d19d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>no_KMT2C</th>\n",
       "      <th>KMT2C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2389_CL16</td>\n",
       "      <td>0.862730</td>\n",
       "      <td>0.137270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2389_CL5</td>\n",
       "      <td>0.938615</td>\n",
       "      <td>0.061385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2389_CL3</td>\n",
       "      <td>0.927032</td>\n",
       "      <td>0.072968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2389_CL18</td>\n",
       "      <td>0.359055</td>\n",
       "      <td>0.640945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2389_CL14</td>\n",
       "      <td>0.916851</td>\n",
       "      <td>0.083149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2389_CL15</td>\n",
       "      <td>0.758178</td>\n",
       "      <td>0.241822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2389_CL1</td>\n",
       "      <td>0.979575</td>\n",
       "      <td>0.020425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2389_CL17</td>\n",
       "      <td>0.834506</td>\n",
       "      <td>0.165494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2389_CL6</td>\n",
       "      <td>0.920656</td>\n",
       "      <td>0.079344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2389_CL7</td>\n",
       "      <td>0.939910</td>\n",
       "      <td>0.060090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2389_CL4</td>\n",
       "      <td>0.967511</td>\n",
       "      <td>0.032489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2389_CL9</td>\n",
       "      <td>0.877180</td>\n",
       "      <td>0.122820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2389_CL13</td>\n",
       "      <td>0.759235</td>\n",
       "      <td>0.240765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2389_CL12</td>\n",
       "      <td>0.891861</td>\n",
       "      <td>0.108139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2389_CL10</td>\n",
       "      <td>0.970180</td>\n",
       "      <td>0.029820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2389_CL11</td>\n",
       "      <td>0.914322</td>\n",
       "      <td>0.085678</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        clone  no_KMT2C     KMT2C\n",
       "0   2389_CL16  0.862730  0.137270\n",
       "1    2389_CL5  0.938615  0.061385\n",
       "2    2389_CL3  0.927032  0.072968\n",
       "3   2389_CL18  0.359055  0.640945\n",
       "4   2389_CL14  0.916851  0.083149\n",
       "5   2389_CL15  0.758178  0.241822\n",
       "6    2389_CL1  0.979575  0.020425\n",
       "7   2389_CL17  0.834506  0.165494\n",
       "8    2389_CL6  0.920656  0.079344\n",
       "9    2389_CL7  0.939910  0.060090\n",
       "10   2389_CL4  0.967511  0.032489\n",
       "11   2389_CL9  0.877180  0.122820\n",
       "12  2389_CL13  0.759235  0.240765\n",
       "13  2389_CL12  0.891861  0.108139\n",
       "14  2389_CL10  0.970180  0.029820\n",
       "15  2389_CL11  0.914322  0.085678"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_2389_prob_df = pd.DataFrame(P_2389, columns=['no_KMT2C', 'KMT2C'])\n",
    "P_2389_prob_df.insert(0, 'clone', RA_2389_clones)\n",
    "P_2389_prob_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "3f8ade90-6e5f-408e-aaaa-1a90c5303e40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_Alias</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Cluster_Assignment</th>\n",
       "      <th>Allelic_CN_minor</th>\n",
       "      <th>Allelic_CN_major</th>\n",
       "      <th>preDP_ccf_mean</th>\n",
       "      <th>preDP_ccf_CI_low</th>\n",
       "      <th>preDP_ccf_CI_high</th>\n",
       "      <th>clust_ccf_mean</th>\n",
       "      <th>clust_ccf_CI_low</th>\n",
       "      <th>clust_ccf_CI_high</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2389_CL1</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL10</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL11</th>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>...</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL12</th>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>...</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL13</th>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>0</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>...</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL14</th>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL15</th>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>0</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>...</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL16</th>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>...</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL17</th>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>...</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL18</th>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>0</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>...</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL3</th>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>...</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL4</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL5</th>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL6</th>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL7</th>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL9</th>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>...</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Patient_ID  Sample_ID  Sample_Alias  Hugo_Symbol  \\\n",
       "Tumor_Sample_Barcode                                                     \n",
       "2389_CL1                      17         17             0           17   \n",
       "2389_CL10                     22         22             0           22   \n",
       "2389_CL11                     65         65             0           65   \n",
       "2389_CL12                     87         87             0           87   \n",
       "2389_CL13                    206        206             0          206   \n",
       "2389_CL14                     64         64             0           64   \n",
       "2389_CL15                    205        205             0          205   \n",
       "2389_CL16                    112        112             0          112   \n",
       "2389_CL17                    133        133             0          133   \n",
       "2389_CL18                    775        775             0          775   \n",
       "2389_CL3                      57         57             0           57   \n",
       "2389_CL4                      24         24             0           24   \n",
       "2389_CL5                      46         46             0           46   \n",
       "2389_CL6                      60         60             0           60   \n",
       "2389_CL7                      46         46             0           46   \n",
       "2389_CL9                      98         98             0           98   \n",
       "\n",
       "                      Chromosome  Start_position  Reference_Allele  \\\n",
       "Tumor_Sample_Barcode                                                 \n",
       "2389_CL1                      17              17                17   \n",
       "2389_CL10                     22              22                22   \n",
       "2389_CL11                     65              65                65   \n",
       "2389_CL12                     87              87                87   \n",
       "2389_CL13                    206             206               206   \n",
       "2389_CL14                     64              64                64   \n",
       "2389_CL15                    205             205               205   \n",
       "2389_CL16                    112             112               112   \n",
       "2389_CL17                    133             133               133   \n",
       "2389_CL18                    775             775               775   \n",
       "2389_CL3                      57              57                57   \n",
       "2389_CL4                      24              24                24   \n",
       "2389_CL5                      46              46                46   \n",
       "2389_CL6                      60              60                60   \n",
       "2389_CL7                      46              46                46   \n",
       "2389_CL9                      98              98                98   \n",
       "\n",
       "                      Tumor_Seq_Allele  t_ref_count  t_alt_count  ...  \\\n",
       "Tumor_Sample_Barcode                                              ...   \n",
       "2389_CL1                            17           17           17  ...   \n",
       "2389_CL10                           22           22           22  ...   \n",
       "2389_CL11                           65           65           65  ...   \n",
       "2389_CL12                           87           87           87  ...   \n",
       "2389_CL13                          206          206          206  ...   \n",
       "2389_CL14                           64           64           64  ...   \n",
       "2389_CL15                          205          205          205  ...   \n",
       "2389_CL16                          112          112          112  ...   \n",
       "2389_CL17                          133          133          133  ...   \n",
       "2389_CL18                          775          775          775  ...   \n",
       "2389_CL3                            57           57           57  ...   \n",
       "2389_CL4                            24           24           24  ...   \n",
       "2389_CL5                            46           46           46  ...   \n",
       "2389_CL6                            60           60           60  ...   \n",
       "2389_CL7                            46           46           46  ...   \n",
       "2389_CL9                            98           98           98  ...   \n",
       "\n",
       "                      Variant_Type  Cluster_Assignment  Allelic_CN_minor  \\\n",
       "Tumor_Sample_Barcode                                                       \n",
       "2389_CL1                        17                  17                 0   \n",
       "2389_CL10                       22                  22                 0   \n",
       "2389_CL11                       65                  65                 0   \n",
       "2389_CL12                       87                  87                 0   \n",
       "2389_CL13                      206                 206                 0   \n",
       "2389_CL14                       64                  64                 0   \n",
       "2389_CL15                      205                 205                 0   \n",
       "2389_CL16                      112                 112                 0   \n",
       "2389_CL17                      133                 133                 0   \n",
       "2389_CL18                      775                 775                 0   \n",
       "2389_CL3                        57                  57                 0   \n",
       "2389_CL4                        24                  24                 0   \n",
       "2389_CL5                        46                  46                 0   \n",
       "2389_CL6                        60                  60                 0   \n",
       "2389_CL7                        46                  46                 0   \n",
       "2389_CL9                        98                  98                 0   \n",
       "\n",
       "                      Allelic_CN_major  preDP_ccf_mean  preDP_ccf_CI_low  \\\n",
       "Tumor_Sample_Barcode                                                       \n",
       "2389_CL1                             0              17                17   \n",
       "2389_CL10                            0              22                22   \n",
       "2389_CL11                            0              65                65   \n",
       "2389_CL12                            0              87                87   \n",
       "2389_CL13                            0             206               206   \n",
       "2389_CL14                            0              64                64   \n",
       "2389_CL15                            0             205               205   \n",
       "2389_CL16                            0             112               112   \n",
       "2389_CL17                            0             133               133   \n",
       "2389_CL18                            0             775               775   \n",
       "2389_CL3                             0              57                57   \n",
       "2389_CL4                             0              24                24   \n",
       "2389_CL5                             0              46                46   \n",
       "2389_CL6                             0              60                60   \n",
       "2389_CL7                             0              46                46   \n",
       "2389_CL9                             0              98                98   \n",
       "\n",
       "                      preDP_ccf_CI_high  clust_ccf_mean  clust_ccf_CI_low  \\\n",
       "Tumor_Sample_Barcode                                                        \n",
       "2389_CL1                             17              17                17   \n",
       "2389_CL10                            22              22                22   \n",
       "2389_CL11                            65              65                65   \n",
       "2389_CL12                            87              87                87   \n",
       "2389_CL13                           206             206               206   \n",
       "2389_CL14                            64              64                64   \n",
       "2389_CL15                           205             205               205   \n",
       "2389_CL16                           112             112               112   \n",
       "2389_CL17                           133             133               133   \n",
       "2389_CL18                           775             775               775   \n",
       "2389_CL3                             57              57                57   \n",
       "2389_CL4                             24              24                24   \n",
       "2389_CL5                             46              46                46   \n",
       "2389_CL6                             60              60                60   \n",
       "2389_CL7                             46              46                46   \n",
       "2389_CL9                             98              98                98   \n",
       "\n",
       "                      clust_ccf_CI_high  \n",
       "Tumor_Sample_Barcode                     \n",
       "2389_CL1                             17  \n",
       "2389_CL10                            22  \n",
       "2389_CL11                            65  \n",
       "2389_CL12                            87  \n",
       "2389_CL13                           206  \n",
       "2389_CL14                            64  \n",
       "2389_CL15                           205  \n",
       "2389_CL16                           112  \n",
       "2389_CL17                           133  \n",
       "2389_CL18                           775  \n",
       "2389_CL3                             57  \n",
       "2389_CL4                             24  \n",
       "2389_CL5                             46  \n",
       "2389_CL6                             60  \n",
       "2389_CL7                             46  \n",
       "2389_CL9                             98  \n",
       "\n",
       "[16 rows x 22 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RA_2389.groupby('Tumor_Sample_Barcode').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4caa1ec-5c59-449c-b283-a345e20a6959",
   "metadata": {},
   "source": [
    "## patient score if no KMT2C mutation in any clone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6af28654-348a-4d47-b470-13b92097a7a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.6946982684320844\n"
     ]
    }
   ],
   "source": [
    "P_2389_prob_df_no_KMT2C = P_2389_prob_df['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C = [ np.log(p) for p in P_2389_prob_df_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8fab121-cf22-4aa9-8d20-1ff5a407d6af",
   "metadata": {},
   "source": [
    "## patient score if a KMT2C mutation in all clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e3252520-4456-4413-9daa-4eb7ed82fae5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37.92316367280446\n"
     ]
    }
   ],
   "source": [
    "P_2389_prob_df_KMT2C = P_2389_prob_df['KMT2C'].to_list()\n",
    "logs_KMT2C = [ np.log(p) for p in P_2389_prob_df_KMT2C]\n",
    "print( -np.sum(logs_KMT2C))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29c85c9e-7529-44bf-a898-04086f2c2398",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_score_real():\n",
    "    probs_pats_KMT2C_scores = []\n",
    "\n",
    "    all_clones_in_pat = list(pat_probs['clone'])\n",
    "    for clone in all_clones_in_pat:\n",
    "        pat_probs_clone = pat_probs[pat_probs['clone']==clone]\n",
    "        if clone in all_KMT2C_clones:\n",
    "\n",
    "            pat_probs_clone_KMT2C = pat_probs_clone['KMT2C'].to_list()\n",
    "            logs_KMT2C = [ - np.log(p) for p in pat_probs_clone_KMT2C]\n",
    "            running_score.extend(logs_KMT2C)\n",
    "        else:\n",
    "            pat_probs_no_KMT2C = pat_probs_clone['no_KMT2C'].to_list()\n",
    "            logs_no_KMT2C = [ - np.log(p) for p in pat_probs_no_KMT2C]\n",
    "            running_score.extend(logs_no_KMT2C)\n",
    "            \n",
    "    score = np.sum(running_score)\n",
    "        \n",
    "    probs_pats_KMT2C_scores.append(score)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3f7ffe8-e462-41d2-8269-e10dc282bbb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_score_perms():\n",
    "\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26290a7e-ef24-4395-aaf8-e65e15c0000d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8b399b63-c359-4018-917b-c7844e54eb42",
   "metadata": {},
   "source": [
    "## permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9350e0c-df83-4898-aaf7-353c2261bad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## so I am using the curveball to continue to permute the number of mutations per clone given the number of mutations and clones in the data\n",
    "## again start on a per patient basis\n",
    "\n",
    "## OK so my distribution is the probability matrix that I get from the observed\n",
    "## then through the curveball, I permute # of clones mutated and calculate the score\n",
    "## then compare the real score to the observed score under the perms\n",
    "\n",
    "## in this case it is also important to know which clones I am permuting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375a7c2c-ce5a-4440-8c2d-31defd1124c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "maf = coding_subclones_maf\n",
    "clones, c_ui, c_idx = np.unique(maf[\"Tumor_Sample_Barcode\"], return_inverse = True, return_index = True)\n",
    "\n",
    "_, p_idx = np.unique(maf[\"Patient_ID\"], return_inverse = True)\n",
    "c2pat = p_idx[c_ui]\n",
    "\n",
    "# genes\n",
    "genes, g_ui, g_idx = np.unique(maf[\"Hugo_Symbol\"], return_inverse = True, return_index = True)\n",
    "g2gidx = dict(zip(genes, g_idx[g_ui]))\n",
    "\n",
    "subset_idx_mat = [g2gidx[x] for x in S.loc[subset_idx, \"gene\"]]\n",
    "\n",
    "# CG is matrix size # total of genes x total # of clones\n",
    "CG = np.array((sp.coo_matrix((np.ones_like(c_idx), (c_idx, g_idx))).todense() > 0))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "66e486c1-8e79-428d-9b76-0d2d6b02a391",
   "metadata": {},
   "outputs": [],
   "source": [
    "RA2389_obs_score = probs_pats_with_KMT2C_scores_dict[2389]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee7e6bd2-9219-41ae-b601-a793463676b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# precompute nonzero elements\n",
    "hp = curveball.find_presences(CG)\n",
    "\n",
    "#\n",
    "# run permutations (in parallel)\n",
    "pool = multiprocessing.Pool()\n",
    "\n",
    "n_perm = 10000\n",
    "n_chunk = 1000\n",
    "\n",
    "n_genes=1\n",
    "n_iterations = n_perm // n_chunk \n",
    "total_clones = np.arange(1, len(RA_2389_clones) )\n",
    "mutation_counts = np.zeros((n_genes, len(RA_2389_clones)))\n",
    "test_matrix = np.zeros((n_iterations, n_genes))\n",
    "\n",
    "test = np.zeros_like(null_pat_sum)\n",
    "for p_chunk in range(n_perm//n_chunk):\n",
    "    permute_futures = [None]*n_chunk\n",
    "    for p in range(n_chunk):\n",
    "        \n",
    "        permute_futures[p] = pool.apply_async(curveball.curve_ball, (CG, hp, 1000) )\n",
    "        \n",
    "        #permute_futures[p].get() = size of total # of clones\n",
    "\n",
    "    # compute permutation p-values\n",
    "    for p in range(n_chunk):\n",
    "\n",
    "        # comput_pat_sums(total # of clones(get cell of matrix[:, genes we are testing (14)] <= observed distribution)\n",
    "        \n",
    "        # returns size 14 matrix because 14 genes, each entry is the # of patients where that gene was mutated\n",
    "        itr_test = ( compute_pat_sums(c2pat, permute_futures[p].get()[:, subset_idx_mat]))\n",
    "        \n",
    "        # c2pat maps clones to patients\n",
    "        test += ( compute_pat_sums(c2pat, permute_futures[p].get()[:, subset_idx_mat]) <= RA2389_obs_score)\n",
    "\n",
    "        for gene_idx, num_mut_patients in enumerate(itr_test):\n",
    "            if 1 <= num_mut_patients <= 15:  \n",
    "                mutation_counts[gene_idx, num_mut_patients - 1] += 1\n",
    "    \n",
    "    # tracking the final deviation from the observed distribution of # patients with a given mutation on each step         \n",
    "    test_matrix[p_chunk, :] = itr_test.copy()\n",
    "    \n",
    "    print(f\"Chunk {p_chunk + 1}/{n_perm//n_chunk} finished.\")\n",
    "\n",
    "#Cb = pd.DataFrame({ \"gene\" : genes[subset_idx_mat], \"p\" : test/(n_chunk*(n_perm//n_chunk)) })\n",
    "\n",
    "Cb = pd.DataFrame({ \"gene\" : genes[subset_idx_mat], \"p\" : test/(n_chunk*(n_perm//n_chunk)),\n",
    "                      \"n_mut_clones\": n_mut_clones[subset_idx_mat], \n",
    "                    \"n_mut_pats\": n_mut_pats[subset_idx_mat] })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e921803-bd9a-4297-850c-906dd2d6945b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e9c27f4-cf46-4c39-a692-72d94671f243",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6001fbd4-051c-420f-a012-fe0916f5537e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "32568f76-ad33-4f8a-aecf-758a986c41e3",
   "metadata": {},
   "source": [
    "## test RA 1598, no KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ff0149b2-f0c5-4ab7-adb7-1e4c9f52f7e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "RA_1598= coding_maf[coding_maf['Patient_ID']==1598]\n",
    "RA_1598_clones = list(RA_1598['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "# total number of mutations in this patient\n",
    "M = len(RA_1598)\n",
    "# initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "P_1598 =np.zeros( (len(RA_1598_clones), 2 ))\n",
    "\n",
    "gene=\"KMT2C\"\n",
    "\n",
    "clone_idx=0\n",
    "for clone in RA_1598_clones:\n",
    "    clone_df = RA_1598[RA_1598['Tumor_Sample_Barcode'] == clone]\n",
    "    mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "    Mij = len(list(mutations_in_clone))\n",
    "    term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "    # a given clone has no mutation in the gene\n",
    "    P_1598[clone_idx, 0] =  term_to_calc\n",
    "    \n",
    "    #  a given clone has a mutation in the gene\n",
    "    P_1598[clone_idx, 1] = 1-term_to_calc\n",
    "    \n",
    "\n",
    "    clone_idx +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b8233f8a-9598-4d47-a4a0-6f3d4c550d79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>no_KMT2C</th>\n",
       "      <th>KMT2C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1598_CL2</td>\n",
       "      <td>0.931663</td>\n",
       "      <td>0.068337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1598_CL1</td>\n",
       "      <td>0.867789</td>\n",
       "      <td>0.132211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1598_CL8</td>\n",
       "      <td>0.700242</td>\n",
       "      <td>0.299758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1598_CL10</td>\n",
       "      <td>0.269250</td>\n",
       "      <td>0.730750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1598_CL5</td>\n",
       "      <td>0.883355</td>\n",
       "      <td>0.116645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1598_CL4</td>\n",
       "      <td>0.931663</td>\n",
       "      <td>0.068337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1598_CL7</td>\n",
       "      <td>0.793799</td>\n",
       "      <td>0.206201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1598_CL9</td>\n",
       "      <td>0.867789</td>\n",
       "      <td>0.132211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1598_CL3</td>\n",
       "      <td>0.899186</td>\n",
       "      <td>0.100814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1598_CL6</td>\n",
       "      <td>0.793799</td>\n",
       "      <td>0.206201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       clone  no_KMT2C     KMT2C\n",
       "0   1598_CL2  0.931663  0.068337\n",
       "1   1598_CL1  0.867789  0.132211\n",
       "2   1598_CL8  0.700242  0.299758\n",
       "3  1598_CL10  0.269250  0.730750\n",
       "4   1598_CL5  0.883355  0.116645\n",
       "5   1598_CL4  0.931663  0.068337\n",
       "6   1598_CL7  0.793799  0.206201\n",
       "7   1598_CL9  0.867789  0.132211\n",
       "8   1598_CL3  0.899186  0.100814\n",
       "9   1598_CL6  0.793799  0.206201"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_1598_prob_df = pd.DataFrame(P_1598, columns=['no_KMT2C', 'KMT2C'])\n",
    "P_1598_prob_df.insert(0, 'clone', RA_1598_clones)\n",
    "P_1598_prob_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16242eda-896b-4dc8-aee1-1f438e56f3a8",
   "metadata": {},
   "source": [
    "## patient score if no KMT2C clones (truth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "abb3dc3b-4b06-4265-9d51-a92e05ee103d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.785767361195695\n"
     ]
    }
   ],
   "source": [
    "P_1598_prob_df_no_KMT2C = P_1598_prob_df['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C_1598 = [ np.log(p) for p in P_1598_prob_df_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C_1598))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ccf5843-6c34-4980-a4c8-cfb48e4a1b0c",
   "metadata": {},
   "source": [
    "## patient score if all KMT2C clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3465bf8e-6f37-4fb6-9d6f-fa7bdd57feef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18.532696469275578\n"
     ]
    }
   ],
   "source": [
    "P_1598_prob_df_KMT2C = P_1598_prob_df['KMT2C'].to_list()\n",
    "logs_KMT2C_1598 = [ np.log(p) for p in P_1598_prob_df_KMT2C]\n",
    "print( -np.sum(logs_KMT2C_1598))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7e110c6-a2d7-4e39-97ee-2aa317bd4c88",
   "metadata": {},
   "source": [
    "## try for all patients for KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "25f06c03-9d72-40cc-815c-b4816af8e0a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_patients = list(combined_maf['Patient_ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "4c0f1cb5-7e75-4c54-88f3-17209c400de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_probs_KMT2C = []\n",
    "\n",
    "for patient in all_patients: \n",
    "\n",
    "    RA_maf = coding_maf[coding_maf['Patient_ID']==patient]\n",
    "    RA_maf_clones = list(RA_maf['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "    ## total number of mutations in the target gene N0\n",
    "    N0 = len(KMT2C_all_muts)\n",
    "    ## total number of clones\n",
    "    N_c = len( coding_maf['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "    # total number of mutations in this patient\n",
    "    M = len(RA_maf)\n",
    "    # initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "    P_pat =np.zeros( (len(RA_maf_clones), 2 ))\n",
    "    \n",
    "    gene=\"KMT2C\"\n",
    "    \n",
    "    clone_idx=0\n",
    "    for clone in RA_maf_clones:\n",
    "        clone_df = RA_maf[RA_maf['Tumor_Sample_Barcode'] == clone]\n",
    "        mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "        Mij = len(list(mutations_in_clone))\n",
    "        \n",
    "        term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "        \n",
    "        # probability that a given clone has no mutation in the gene\n",
    "        P_pat[clone_idx, 0] = term_to_calc\n",
    "        \n",
    "        # probability that a given clone has one or more mutations in the gene\n",
    "        P_pat[clone_idx, 1] = 1- term_to_calc\n",
    "        clone_idx +=1\n",
    "\n",
    "    P_prob_df = pd.DataFrame(P_pat, columns=['no_KMT2C', 'KMT2C'])\n",
    "    P_prob_df.insert(0, 'clone', RA_maf_clones)\n",
    "    all_probs_KMT2C.append(P_prob_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a4682501-36e8-4aab-9f2d-94bd54121813",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs_1326 = all_probs_KMT2C[11]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8ac75a0-00cf-464a-befe-8f6dbf6d9022",
   "metadata": {},
   "source": [
    "## score if no KMT2C mutation in this patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6b9ae65e-4bad-4a09-81c3-f66e005cd019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.785239956066282\n"
     ]
    }
   ],
   "source": [
    "probs_1326_no_KMT2C = probs_1326['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C_1326 = [ np.log(p) for p in probs_1326_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C_1326))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a69120-e34b-472d-bc14-4a6bcbb142d9",
   "metadata": {},
   "source": [
    "## real distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4ce52f65-5230-4dd1-8915-f1fb2020a9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "KMT2C_clones_RA1326= ['1326_CL2','1326_CL9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "39236529-1bb1-481e-b831-d50c9231b90c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.547794414031647"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_1326_prob_df_KMT2C_clones = probs_1326[probs_1326['clone'].isin(KMT2C_clones_RA1326)]\n",
    "P_1326_prob_df_non_KMT2C_clones = probs_1326[~probs_1326['clone'].isin(KMT2C_clones_RA1326)]\n",
    "\n",
    "P_1326_prob_df_KMT2C_clones_list = P_1326_prob_df_KMT2C_clones['KMT2C'].to_list()\n",
    "logs_real_KMT2C = [ -np.log(p) for p in P_1326_prob_df_KMT2C_clones_list]\n",
    "\n",
    "P_1326_no_KMT2C_clones_list = P_1326_prob_df_non_KMT2C_clones['no_KMT2C'].to_list()\n",
    "logs_real_no_KMT2C = [ -np.log(p) for p in P_1326_no_KMT2C_clones_list]\n",
    "\n",
    "\n",
    "np.sum( np.concatenate((logs_real_KMT2C, logs_real_no_KMT2C)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b374f593-d745-4049-9114-f673aba9215d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[        clone  no_KMT2C     KMT2C\n",
       " 0    1547_CL1  0.480690  0.519310\n",
       " 1    1547_CL8  0.899154  0.100846\n",
       " 2    1547_CL2  0.894124  0.105876\n",
       " 3   1547_CL14  0.297655  0.702345\n",
       " 4   1547_CL15  0.709490  0.290510\n",
       " 5   1547_CL12  0.919544  0.080456\n",
       " 6    1547_CL4  0.994434  0.005566\n",
       " 7    1547_CL6  0.988898  0.011102\n",
       " 8    1547_CL3  0.988898  0.011102\n",
       " 9   1547_CL11  0.864503  0.135497\n",
       " 10  1547_CL13  0.994434  0.005566,\n",
       "        clone  no_KMT2C     KMT2C\n",
       " 0   1598_CL2  0.931663  0.068337\n",
       " 1   1598_CL1  0.867789  0.132211\n",
       " 2   1598_CL8  0.700242  0.299758\n",
       " 3  1598_CL10  0.269250  0.730750\n",
       " 4   1598_CL5  0.883355  0.116645\n",
       " 5   1598_CL4  0.931663  0.068337\n",
       " 6   1598_CL7  0.793799  0.206201\n",
       " 7   1598_CL9  0.867789  0.132211\n",
       " 8   1598_CL3  0.899186  0.100814\n",
       " 9   1598_CL6  0.793799  0.206201,\n",
       "       clone  no_KMT2C     KMT2C\n",
       " 0  1078_CL1  0.459814  0.540186\n",
       " 1  1078_CL5  0.590613  0.409387\n",
       " 2  1078_CL2  0.840239  0.159761\n",
       " 3  1078_CL6  0.266166  0.733834\n",
       " 4  1078_CL4  0.965897  0.034103,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0    1045_CL9  0.874180  0.125820\n",
       " 1    1045_CL1  0.847366  0.152634\n",
       " 2    1045_CL3  0.847366  0.152634\n",
       " 3   1045_CL16  0.370728  0.629272\n",
       " 4    1045_CL4  0.949693  0.050307\n",
       " 5   1045_CL10  0.901801  0.098199\n",
       " 6   1045_CL14  0.939925  0.060075\n",
       " 7   1045_CL13  0.618950  0.381050\n",
       " 8   1045_CL15  0.732258  0.267742\n",
       " 9    1045_CL7  0.959557  0.040443\n",
       " 10  1045_CL11  0.796069  0.203931\n",
       " 11   1045_CL6  0.989740  0.010260\n",
       " 12  1045_CL12  0.959557  0.040443,\n",
       "        clone  no_KMT2C     KMT2C\n",
       " 0  1113_CL13  0.942617  0.057383\n",
       " 1   1113_CL1  0.533148  0.466852\n",
       " 2   1113_CL8  0.765482  0.234518\n",
       " 3   1113_CL3  0.837120  0.162880\n",
       " 4   1113_CL2  0.765482  0.234518\n",
       " 5   1113_CL5  0.888378  0.111622\n",
       " 6  1113_CL12  0.639337  0.360663\n",
       " 7  1113_CL11  0.517171  0.482829\n",
       " 8  1113_CL10  0.888378  0.111622,\n",
       "       clone  no_KMT2C     KMT2C\n",
       " 0  1644_CL3  0.490492  0.509508\n",
       " 1  1644_CL1  0.264483  0.735517\n",
       " 2  1644_CL8  0.798251  0.201749\n",
       " 3  1644_CL6  0.644809  0.355191\n",
       " 4  1644_CL7  0.945438  0.054562,\n",
       "        clone  no_KMT2C     KMT2C\n",
       " 0   2819_CL6  0.951902  0.048098\n",
       " 1   2819_CL1  0.806990  0.193010\n",
       " 2  2819_CL10  0.306836  0.693164\n",
       " 3   2819_CL4  0.967685  0.032315\n",
       " 4  2819_CL11  0.780648  0.219352\n",
       " 5   2819_CL8  0.848095  0.151905\n",
       " 6   2819_CL5  0.742666  0.257334\n",
       " 7   2819_CL9  0.607614  0.392386\n",
       " 8   2819_CL7  0.951902  0.048098\n",
       " 9  2819_CL12  0.975669  0.024331,\n",
       "       clone  no_KMT2C     KMT2C\n",
       " 0  2974_CL2  0.784603  0.215397\n",
       " 1  2974_CL8  0.396736  0.603264\n",
       " 2  2974_CL1  0.623400  0.376600\n",
       " 3  2974_CL4  0.913348  0.086652\n",
       " 4  2974_CL9  0.509692  0.490308\n",
       " 5  2974_CL6  0.955739  0.044261\n",
       " 6  2974_CL3  0.955739  0.044261\n",
       " 7  2974_CL5  0.833877  0.166123,\n",
       "        clone  no_KMT2C     KMT2C\n",
       " 0   1035_CL1  0.497882  0.502118\n",
       " 1   1035_CL5  0.763271  0.236729\n",
       " 2   1035_CL6  0.740542  0.259458\n",
       " 3   1035_CL3  0.942017  0.057983\n",
       " 4   1035_CL7  0.810739  0.189261\n",
       " 5   1035_CL4  0.970596  0.029404\n",
       " 6   1035_CL2  0.970596  0.029404\n",
       " 7  1035_CL10  0.546017  0.453983\n",
       " 8   1035_CL9  0.655908  0.344092\n",
       " 9   1035_CL8  0.887245  0.112755,\n",
       "       clone  no_KMT2C     KMT2C\n",
       " 0  1534_CL1  0.107824  0.892176\n",
       " 1  1534_CL4  0.701042  0.298958\n",
       " 2  1534_CL2  0.801486  0.198514\n",
       " 3  1534_CL3  0.915545  0.084455,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0    1558_CL1  0.744842  0.255158\n",
       " 1    1558_CL5  0.967995  0.032005\n",
       " 2    1558_CL6  0.926837  0.073163\n",
       " 3   1558_CL11  0.381214  0.618786\n",
       " 4    1558_CL4  0.947203  0.052797\n",
       " 5    1558_CL3  0.936967  0.063033\n",
       " 6    1558_CL2  0.947203  0.052797\n",
       " 7    1558_CL9  0.858790  0.141210\n",
       " 8    1558_CL8  0.906888  0.093112\n",
       " 9   1558_CL10  0.398890  0.601110\n",
       " 10   1558_CL7  0.967995  0.032005,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0   1326_CL13  0.362515  0.637485\n",
       " 1    1326_CL6  0.840239  0.159761\n",
       " 2    1326_CL1  0.779492  0.220508\n",
       " 3   1326_CL10  0.905479  0.094521\n",
       " 4    1326_CL2  0.741343  0.258657\n",
       " 5    1326_CL3  0.975526  0.024474\n",
       " 6    1326_CL7  0.928278  0.071722\n",
       " 7    1326_CL9  0.670312  0.329688\n",
       " 8    1326_CL5  0.975526  0.024474\n",
       " 9   1326_CL12  0.670312  0.329688\n",
       " 10   1326_CL8  0.975526  0.024474,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0   2389_CL16  0.862730  0.137270\n",
       " 1    2389_CL5  0.938615  0.061385\n",
       " 2    2389_CL3  0.927032  0.072968\n",
       " 3   2389_CL18  0.359055  0.640945\n",
       " 4   2389_CL14  0.916851  0.083149\n",
       " 5   2389_CL15  0.758178  0.241822\n",
       " 6    2389_CL1  0.979575  0.020425\n",
       " 7   2389_CL17  0.834506  0.165494\n",
       " 8    2389_CL6  0.920656  0.079344\n",
       " 9    2389_CL7  0.939910  0.060090\n",
       " 10   2389_CL4  0.967511  0.032489\n",
       " 11   2389_CL9  0.877180  0.122820\n",
       " 12  2389_CL13  0.759235  0.240765\n",
       " 13  2389_CL12  0.891861  0.108139\n",
       " 14  2389_CL10  0.970180  0.029820\n",
       " 15  2389_CL11  0.914322  0.085678,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0   2542_CL12  0.354363  0.645637\n",
       " 1   2542_CL10  0.990646  0.009354\n",
       " 2    2542_CL1  0.859212  0.140788\n",
       " 3    2542_CL3  0.928257  0.071743\n",
       " 4   2542_CL18  0.793018  0.206982\n",
       " 5   2542_CL16  0.614770  0.385230\n",
       " 6   2542_CL15  0.832270  0.167730\n",
       " 7    2542_CL2  0.978861  0.021139\n",
       " 8   2542_CL17  0.925872  0.074128\n",
       " 9    2542_CL9  0.987265  0.012735\n",
       " 10   2542_CL7  0.989799  0.010201\n",
       " 11  2542_CL13  0.954055  0.045945\n",
       " 12   2542_CL6  0.968866  0.031134\n",
       " 13   2542_CL5  0.999146  0.000854\n",
       " 14  2542_CL14  0.884700  0.115300\n",
       " 15  2542_CL11  0.992340  0.007660\n",
       " 16  2542_CL19  0.883181  0.116819\n",
       " 17   2542_CL4  0.991493  0.008507\n",
       " 18   2542_CL8  0.999146  0.000854,\n",
       "         clone  no_KMT2C     KMT2C\n",
       " 0    1002_CL1  0.961869  0.038131\n",
       " 1   1002_CL21  0.203368  0.796632\n",
       " 2   1002_CL11  0.991317  0.008683\n",
       " 3    1002_CL3  0.987393  0.012607\n",
       " 4    1002_CL8  0.971077  0.028923\n",
       " 5   1002_CL10  0.872157  0.127843\n",
       " 6   1002_CL20  0.751477  0.248523\n",
       " 7   1002_CL19  0.928073  0.071927\n",
       " 8   1002_CL16  0.933255  0.066745\n",
       " 9    1002_CL6  0.995255  0.004745\n",
       " 10  1002_CL18  0.955017  0.044983\n",
       " 11  1002_CL15  0.971848  0.028152\n",
       " 12  1002_CL13  0.971077  0.028923\n",
       " 13  1002_CL12  0.932513  0.067487\n",
       " 14   1002_CL5  0.992890  0.007110\n",
       " 15  1002_CL17  0.897554  0.102446\n",
       " 16  1002_CL14  0.978814  0.021186]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_probs_KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "605b99ce-4506-4f86-9d1f-8e9ec58a6ec5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_probs_order = [ 1547, 1598, 1078, 1045, 1113, 1644, 2819, 2974, 1035,1534, 1558, 1326, 2389, 2542, 1002  ]\n",
    "len(all_probs_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "30603c03-8872-4ba5-83f8-b52293d4581c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pats_with_no_KMT2C = [1598, 1078, 1045, 1113, 1644, 2819, 2974, 1035,1534,1002]\n",
    "pats_with_no_KMT2C_idx =[1, 2, 3, 4, 5, 6, 7, 8, 9, 14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b8956df6-9b6b-412a-b0a0-e71d74e31b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs_pats_no_KMT2C = []\n",
    "probs_pats_no_KMT2C_scores = []\n",
    "probs_pats_no_KMT2C_scores_dict = {}\n",
    "for idx in pats_with_no_KMT2C_idx:\n",
    "    score = 0\n",
    "    probs_pats_no_KMT2C.append(all_probs_KMT2C[idx])\n",
    "    pat_probs = all_probs_KMT2C[idx]\n",
    "    pat_probs_no_KMT2C = pat_probs['no_KMT2C'].to_list()\n",
    "    logs_no_KMT2C = [ np.log(p) for p in pat_probs_no_KMT2C]\n",
    "    score = -np.sum(logs_no_KMT2C)\n",
    "    \n",
    "    probs_pats_no_KMT2C_scores.append(score)\n",
    "    pat = all_probs_order[idx]\n",
    "    probs_pats_no_KMT2C_scores_dict[pat] = score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f872a4f5-6500-4467-bc1c-7454eaa81afa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1598: 2.785767361195695,\n",
       " 1078: 2.8359286696937476,\n",
       " 1045: 2.787239669333773,\n",
       " 1113: 2.743758971093486,\n",
       " 1644: 2.762563428360604,\n",
       " 2819: 2.760069539439102,\n",
       " 2974: 2.676423530556352,\n",
       " 1035: 2.7436121458661056,\n",
       " 1534: 2.891962217736308,\n",
       " 1002: 2.56378181686393}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs_pats_no_KMT2C_scores_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a0ed914d-ec21-466d-bd21-e57ed1156d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "pats_with_KMT2C_idx=[0, 10,11,12,13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "708331f5-7dad-4ee2-bbc3-13cad03247ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_KMT2C_clones = list(KMT2C_all_muts['Tumor_Sample_Barcode'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ab44785a-578a-4620-9321-7b8197f33edc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "probs_pats_with_KMT2C = []\n",
    "probs_pats_KMT2C_scores = []\n",
    "probs_pats_with_KMT2C_scores_dict = {}\n",
    "for idx in pats_with_KMT2C_idx:\n",
    "    #score=0\n",
    "    running_score=[]\n",
    "    pat_probs = all_probs_KMT2C[idx]\n",
    "    all_clones_in_pat = list(pat_probs['clone'])\n",
    "    for clone in all_clones_in_pat:\n",
    "        pat_probs_clone = pat_probs[pat_probs['clone']==clone]\n",
    "        if clone in all_KMT2C_clones:\n",
    "\n",
    "            pat_probs_clone_KMT2C = pat_probs_clone['KMT2C'].to_list()\n",
    "            logs_KMT2C = [ - np.log(p) for p in pat_probs_clone_KMT2C]\n",
    "            running_score.extend(logs_KMT2C)\n",
    "        else:\n",
    "            pat_probs_no_KMT2C = pat_probs_clone['no_KMT2C'].to_list()\n",
    "            logs_no_KMT2C = [ - np.log(p) for p in pat_probs_no_KMT2C]\n",
    "            running_score.extend(logs_no_KMT2C)\n",
    "            \n",
    "    score = np.sum(running_score)\n",
    "    \n",
    "    probs_pats_with_KMT2C.append(pat_probs)\n",
    "    probs_pats_KMT2C_scores.append(score)\n",
    "    pat = all_probs_order[idx]\n",
    "    probs_pats_with_KMT2C_scores_dict[pat] = score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "458ad730-0f9e-4ba0-bd9c-fcedca07be6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1547: 2.8031613827454733,\n",
       " 1558: 3.403834833843852,\n",
       " 1326: 4.547794414031647,\n",
       " 2389: 14.714030365894862,\n",
       " 2542: 8.820179860729803}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs_pats_with_KMT2C_scores_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f33186-5ad8-4705-8964-715c39980a16",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41d08f18-d731-4224-b6a3-039f3e617eb3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18cc4439-53d7-48f2-87fa-7053a21739b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0af8a38a-bab1-4852-885b-23145742a166",
   "metadata": {},
   "source": [
    "## test any gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "9969e7cb-1f17-49dd-ad95-741ac1b55640",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_distributions(gene):\n",
    "\n",
    "    gene_all_muts = coding_maf[coding_maf['Hugo_Symbol']==gene]\n",
    "    all_probs = []\n",
    "    \n",
    "    for patient in all_patients: \n",
    "    \n",
    "        RA_maf = coding_maf[coding_maf['Patient_ID']==patient]\n",
    "        RA_maf_clones = list(RA_maf['Tumor_Sample_Barcode'].unique() )\n",
    "        \n",
    "        ## total number of mutations N0\n",
    "        N0 = len(gene_all_muts)\n",
    "        ## total number of mutated clones N_c\n",
    "        N_c = len( coding_maf['Tumor_Sample_Barcode'].unique() )\n",
    "        \n",
    "        # total number of mutations in this patient\n",
    "        M = len(RA_maf)\n",
    "        # initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "        P_pat =np.zeros( (len(RA_maf_clones), 2 ))\n",
    "        \n",
    "        clone_idx=0\n",
    "        for clone in RA_maf_clones:\n",
    "            clone_df = RA_maf[RA_maf['Tumor_Sample_Barcode'] == clone]\n",
    "            mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "            Mij = len(list(mutations_in_clone))\n",
    "            term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "\n",
    "            # probability that a given clone has no mutation in the gene\n",
    "            P_pat[clone_idx, 0] = term_to_calc\n",
    "            \n",
    "            \n",
    "            # probability that a given clone has a mutation in the gene\n",
    "            P_pat[clone_idx, 1] = 1-term_to_calc\n",
    "            clone_idx +=1\n",
    "\n",
    "    \n",
    "        P_prob_df = pd.DataFrame(P_pat, columns=['no_gene_mutations', 'gene_mutation'])\n",
    "        P_prob_df.insert(0, 'clone', RA_maf_clones)\n",
    "        all_probs.append(P_prob_df)\n",
    "    return all_probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "65f550ed-4630-47af-9f0f-900274af9576",
   "metadata": {},
   "outputs": [],
   "source": [
    "esr1_Pm = calculate_distributions('ESR1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "812883a7-f186-4944-a61f-128cafc538bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[        clone  no_gene_mutations  gene_mutation\n",
       " 0    1547_CL1           0.848241       0.151759\n",
       " 1    1547_CL8           0.976215       0.023785\n",
       " 2    1547_CL2           0.974977       0.025023\n",
       " 3   1547_CL14           0.762891       0.237109\n",
       " 4   1547_CL15           0.925434       0.074566\n",
       " 5   1547_CL12           0.981180       0.018820\n",
       " 6    1547_CL4           0.998735       0.001265\n",
       " 7    1547_CL6           0.997472       0.002528\n",
       " 8    1547_CL3           0.997472       0.002528\n",
       " 9   1547_CL11           0.967581       0.032419\n",
       " 10  1547_CL13           0.998735       0.001265,\n",
       "        clone  no_gene_mutations  gene_mutation\n",
       " 0   1598_CL2           0.984091       0.015909\n",
       " 1   1598_CL1           0.968411       0.031589\n",
       " 2   1598_CL8           0.922709       0.077291\n",
       " 3  1598_CL10           0.746270       0.253730\n",
       " 4   1598_CL5           0.972310       0.027690\n",
       " 5   1598_CL4           0.984091       0.015909\n",
       " 6   1598_CL7           0.949127       0.050873\n",
       " 7   1598_CL9           0.968411       0.031589\n",
       " 8   1598_CL3           0.976223       0.023777\n",
       " 9   1598_CL6           0.949127       0.050873,\n",
       "       clone  no_gene_mutations  gene_mutation\n",
       " 0  1078_CL1           0.839903       0.160097\n",
       " 1  1078_CL5           0.888140       0.111860\n",
       " 2  1078_CL2           0.961380       0.038620\n",
       " 3  1078_CL6           0.744386       0.255614\n",
       " 4  1078_CL4           0.992166       0.007834,\n",
       "         clone  no_gene_mutations  gene_mutation\n",
       " 0    1045_CL9           0.970018       0.029982\n",
       " 1    1045_CL1           0.963215       0.036785\n",
       " 2    1045_CL3           0.963215       0.036785\n",
       " 3   1045_CL16           0.800734       0.199266\n",
       " 4    1045_CL4           0.988371       0.011629\n",
       " 5   1045_CL10           0.976864       0.023136\n",
       " 6   1045_CL14           0.986060       0.013940\n",
       " 7   1045_CL13           0.897509       0.102491\n",
       " 8   1045_CL15           0.932032       0.067968\n",
       " 9    1045_CL7           0.990687       0.009313\n",
       " 10  1045_CL11           0.949738       0.050262\n",
       " 11   1045_CL6           0.997664       0.002336\n",
       " 12  1045_CL12           0.990687       0.009313,\n",
       "        clone  no_gene_mutations  gene_mutation\n",
       " 0  1113_CL13           0.986699       0.013301\n",
       " 1   1113_CL1           0.868053       0.131947\n",
       " 2   1113_CL8           0.941389       0.058611\n",
       " 3   1113_CL3           0.960573       0.039427\n",
       " 4   1113_CL2           0.941389       0.058611\n",
       " 5   1113_CL5           0.973557       0.026443\n",
       " 6  1113_CL12           0.904052       0.095948\n",
       " 7  1113_CL11           0.862180       0.137820\n",
       " 8  1113_CL10           0.973557       0.026443,\n",
       "       clone  no_gene_mutations  gene_mutation\n",
       " 0  1644_CL3           0.852062       0.147938\n",
       " 1  1644_CL1           0.743352       0.256648\n",
       " 2  1644_CL8           0.950324       0.049676\n",
       " 3  1644_CL6           0.905782       0.094218\n",
       " 4  1644_CL7           0.987367       0.012633,\n",
       "        clone  no_gene_mutations  gene_mutation\n",
       " 0   2819_CL6           0.988891       0.011109\n",
       " 1   2819_CL1           0.952660       0.047340\n",
       " 2  2819_CL10           0.768006       0.231994\n",
       " 3   2819_CL4           0.992582       0.007418\n",
       " 4  2819_CL11           0.945559       0.054441\n",
       " 5   2819_CL8           0.963402       0.036598\n",
       " 6   2819_CL5           0.934997       0.065003\n",
       " 7   2819_CL9           0.893801       0.106199\n",
       " 8   2819_CL7           0.988891       0.011109\n",
       " 9  2819_CL12           0.994432       0.005568,\n",
       "       clone  no_gene_mutations  gene_mutation\n",
       " 0  2974_CL2           0.946637       0.053363\n",
       " 1  2974_CL8           0.812842       0.187158\n",
       " 2  2974_CL1           0.898951       0.101049\n",
       " 3  2974_CL4           0.979680       0.020320\n",
       " 4  2974_CL9           0.859384       0.140616\n",
       " 5  2974_CL6           0.989793       0.010207\n",
       " 6  2974_CL3           0.989793       0.010207\n",
       " 7  2974_CL5           0.959732       0.040268,\n",
       "        clone  no_gene_mutations  gene_mutation\n",
       " 0   1035_CL1           0.854905       0.145095\n",
       " 1   1035_CL5           0.940776       0.059224\n",
       " 2   1035_CL6           0.934394       0.065606\n",
       " 3   1035_CL3           0.986557       0.013443\n",
       " 4   1035_CL7           0.953657       0.046343\n",
       " 5   1035_CL4           0.993258       0.006742\n",
       " 6   1035_CL2           0.993258       0.006742\n",
       " 7  1035_CL10           0.872688       0.127312\n",
       " 8   1035_CL9           0.909256       0.090744\n",
       " 9   1035_CL8           0.973276       0.026724,\n",
       "       clone  no_gene_mutations  gene_mutation\n",
       " 0  1534_CL1           0.611985       0.388015\n",
       " 1  1534_CL4           0.922946       0.077054\n",
       " 2  1534_CL2           0.951191       0.048809\n",
       " 3  1534_CL3           0.980212       0.019788,\n",
       "         clone  no_gene_mutations  gene_mutation\n",
       " 0    1558_CL1           0.935613       0.064387\n",
       " 1    1558_CL5           0.992654       0.007346\n",
       " 2    1558_CL6           0.982935       0.017065\n",
       " 3   1558_CL11           0.805691       0.194309\n",
       " 4    1558_CL4           0.987784       0.012216\n",
       " 5    1558_CL3           0.985357       0.014643\n",
       " 6    1558_CL2           0.987784       0.012216\n",
       " 7    1558_CL9           0.966133       0.033867\n",
       " 8    1558_CL8           0.978108       0.021892\n",
       " 9   1558_CL10           0.813818       0.186182\n",
       " 10   1558_CL7           0.992654       0.007346,\n",
       "         clone  no_gene_mutations  gene_mutation\n",
       " 0   1326_CL13           0.796778       0.203222\n",
       " 1    1326_CL6           0.961380       0.038620\n",
       " 2    1326_CL1           0.945243       0.054757\n",
       " 3   1326_CL10           0.977764       0.022236\n",
       " 4    1326_CL2           0.934622       0.065378\n",
       " 5    1326_CL3           0.994399       0.005601\n",
       " 6    1326_CL7           0.983281       0.016719\n",
       " 7    1326_CL9           0.913700       0.086300\n",
       " 8    1326_CL5           0.994399       0.005601\n",
       " 9   1326_CL12           0.913700       0.086300\n",
       " 10   1326_CL8           0.994399       0.005601,\n",
       "         clone  no_gene_mutations  gene_mutation\n",
       " 0   2389_CL16           0.967132       0.032868\n",
       " 1    2389_CL5           0.985749       0.014251\n",
       " 2    2389_CL3           0.982982       0.017018\n",
       " 3   2389_CL18           0.795091       0.204909\n",
       " 4   2389_CL14           0.980528       0.019472\n",
       " 5   2389_CL15           0.939358       0.060642\n",
       " 6    2389_CL1           0.995333       0.004667\n",
       " 7   2389_CL17           0.959895       0.040105\n",
       " 8    2389_CL6           0.981448       0.018552\n",
       " 9    2389_CL7           0.986057       0.013943\n",
       " 10   2389_CL4           0.992542       0.007458\n",
       " 11   2389_CL9           0.970769       0.029231\n",
       " 12  2389_CL13           0.939653       0.060347\n",
       " 13  2389_CL12           0.974419       0.025581\n",
       " 14  2389_CL10           0.993161       0.006839\n",
       " 15  2389_CL11           0.979916       0.020084,\n",
       "         clone  no_gene_mutations  gene_mutation\n",
       " 0   2542_CL12           0.792784       0.207216\n",
       " 1   2542_CL10           0.997871       0.002129\n",
       " 2    2542_CL1           0.966240       0.033760\n",
       " 3    2542_CL3           0.983276       0.016724\n",
       " 4   2542_CL18           0.948916       0.051084\n",
       " 5   2542_CL16           0.896148       0.103852\n",
       " 6   2542_CL15           0.959314       0.040686\n",
       " 7    2542_CL2           0.995168       0.004832\n",
       " 8   2542_CL17           0.982703       0.017297\n",
       " 9    2542_CL9           0.997098       0.002902\n",
       " 10   2542_CL7           0.997678       0.002322\n",
       " 11  2542_CL13           0.989398       0.010602\n",
       " 12   2542_CL6           0.992856       0.007144\n",
       " 13   2542_CL5           0.999806       0.000194\n",
       " 14  2542_CL14           0.972644       0.027356\n",
       " 15  2542_CL11           0.998258       0.001742\n",
       " 16  2542_CL19           0.972266       0.027734\n",
       " 17   2542_CL4           0.998065       0.001935\n",
       " 18   2542_CL8           0.999806       0.000194,\n",
       "         clone  no_gene_mutations  gene_mutation\n",
       " 0    1002_CL1           0.991227       0.008773\n",
       " 1   1002_CL21           0.701874       0.298126\n",
       " 2   1002_CL11           0.998025       0.001975\n",
       " 3    1002_CL3           0.997128       0.002872\n",
       " 4    1002_CL8           0.993369       0.006631\n",
       " 5   1002_CL10           0.969510       0.030490\n",
       " 6   1002_CL20           0.937483       0.062517\n",
       " 7   1002_CL19           0.983232       0.016768\n",
       " 8   1002_CL16           0.984472       0.015528\n",
       " 9    1002_CL6           0.998922       0.001078\n",
       " 10  1002_CL18           0.989624       0.010376\n",
       " 11  1002_CL15           0.993548       0.006452\n",
       " 12  1002_CL13           0.993369       0.006631\n",
       " 13  1002_CL12           0.984294       0.015706\n",
       " 14   1002_CL5           0.998383       0.001617\n",
       " 15  1002_CL17           0.975822       0.024178\n",
       " 16  1002_CL14           0.995157       0.004843]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "esr1_Pm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6ceb8c1-338f-4bed-98fe-65d3d1c859eb",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b8beac-0ea2-4568-b94f-c8ecd5ad38b6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2dd7039-5f14-4b9f-b202-a15c199b6c45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "115836c9-0b97-404a-9b98-74479677bf03",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
