{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6c3e7e2a-be98-4a50-aa01-00ae1426fab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv\n",
    "import os\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import scipy.stats as stat\n",
    "import math\n",
    "import scipy.sparse as sp\n",
    "\n",
    "import numpy_groupies as npg\n",
    "import multiprocessing\n",
    "sys.path.append(\"/Users/lmartin/Documents/2023/CDK4_mutsig_added_pats/CurveBall/cbsig\")\n",
    "import curveball"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2793d34-ea60-48b3-b402-bbfa23df3287",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_maf = pd.read_csv('/Users/lmartin/Documents/2025/RA_combined_maf_dedup_updated_on_trees_022025.maf',sep='\\t')\n",
    "subclones_sig_genes = pd.read_csv('/Users/lmartin/Documents/2023/CDK4_mutsig_added_pats/on_subclones_v3/0/mutsig_results/outdir/sig_genes.txt',sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "77258951-8fa7-4eb2-a6ee-9579ea04210c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "subset=(subclones_sig_genes['p'] < 0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98268e6d-d805-4daa-bd20-84322170959d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6790"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_muts=['Missense_Mutation',\n",
    " 'Nonsense_Mutation',\n",
    " 'Start_Codon_SNP',\n",
    " 'De_novo_Start_InFrame',\n",
    " 'De_novo_Start_OutOfFrame',\n",
    " 'Nonstop_Mutation',\n",
    " 'Frame_Shift_Del',\n",
    " 'Frame_Shift_Ins',\n",
    " 'In_Frame_Del',\n",
    " 'In_Frame_Ins']\n",
    "\n",
    "coding_maf=combined_maf[combined_maf['Variant_Classification'].isin(coding_muts)]\n",
    "len(coding_maf['Hugo_Symbol'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d9143db2-80a7-4c3e-abcd-2833aef722d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_mut_pats</th>\n",
       "      <th>n_mut_clones</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CSAD</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNA2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ESR1</th>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H3B</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IVL</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2C</th>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRAS</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEN1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MNDA</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3CA</th>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ST18</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF536</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             n_mut_pats  n_mut_clones\n",
       "Hugo_Symbol                          \n",
       "CSAD                  2             3\n",
       "CTCF                  2             4\n",
       "CTNNA2                5             5\n",
       "ESR1                  6            10\n",
       "HIST1H3B              2             3\n",
       "IVL                   4             6\n",
       "KMT2C                 5            16\n",
       "KRAS                  3             3\n",
       "MEN1                  1             2\n",
       "MNDA                  3             3\n",
       "PIK3CA               11            15\n",
       "ST18                  4             5\n",
       "TP53                  4             6\n",
       "ZNF536                4             5"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([\n",
    "  coding_maf.groupby([\"Hugo_Symbol\", \"Patient_ID\"]).size().loc[subclones_sig_genes.loc[subset, \"gene\"]].groupby(level = 0).size().rename(\"n_mut_pats\"),\n",
    "  coding_maf.groupby([\"Hugo_Symbol\", \"Tumor_Sample_Barcode\"]).size().loc[subclones_sig_genes.loc[subset, \"gene\"]].groupby(level = 0).size().rename(\"n_mut_clones\")],\n",
    "  axis = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b3af026f-cc85-486c-bc71-5aebb0872b88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1002_CL1', '1002_CL10', '1002_CL11', '1002_CL12', '1002_CL13',\n",
       "       '1002_CL14', '1002_CL15', '1002_CL16', '1002_CL17', '1002_CL18',\n",
       "       '1002_CL19', '1002_CL20', '1002_CL21', '1002_CL3', '1002_CL5',\n",
       "       '1002_CL6', '1002_CL8', '1035_CL1', '1035_CL10', '1035_CL2',\n",
       "       '1035_CL3', '1035_CL4', '1035_CL5', '1035_CL6', '1035_CL7',\n",
       "       '1035_CL8', '1035_CL9', '1045_CL1', '1045_CL10', '1045_CL11',\n",
       "       '1045_CL12', '1045_CL13', '1045_CL14', '1045_CL15', '1045_CL16',\n",
       "       '1045_CL3', '1045_CL4', '1045_CL6', '1045_CL7', '1045_CL9',\n",
       "       '1078_CL1', '1078_CL2', '1078_CL4', '1078_CL5', '1078_CL6',\n",
       "       '1113_CL1', '1113_CL10', '1113_CL11', '1113_CL12', '1113_CL13',\n",
       "       '1113_CL2', '1113_CL3', '1113_CL5', '1113_CL8', '1326_CL1',\n",
       "       '1326_CL10', '1326_CL12', '1326_CL13', '1326_CL2', '1326_CL3',\n",
       "       '1326_CL5', '1326_CL6', '1326_CL7', '1326_CL8', '1326_CL9',\n",
       "       '1534_CL1', '1534_CL2', '1534_CL3', '1534_CL4', '1547_CL1',\n",
       "       '1547_CL11', '1547_CL12', '1547_CL13', '1547_CL14', '1547_CL15',\n",
       "       '1547_CL2', '1547_CL3', '1547_CL4', '1547_CL6', '1547_CL8',\n",
       "       '1558_CL1', '1558_CL10', '1558_CL11', '1558_CL2', '1558_CL3',\n",
       "       '1558_CL4', '1558_CL5', '1558_CL6', '1558_CL7', '1558_CL8',\n",
       "       '1558_CL9', '1598_CL1', '1598_CL10', '1598_CL2', '1598_CL3',\n",
       "       '1598_CL4', '1598_CL5', '1598_CL6', '1598_CL7', '1598_CL8',\n",
       "       '1598_CL9', '1644_CL1', '1644_CL3', '1644_CL6', '1644_CL7',\n",
       "       '1644_CL8', '2389_CL1', '2389_CL10', '2389_CL11', '2389_CL12',\n",
       "       '2389_CL13', '2389_CL14', '2389_CL15', '2389_CL16', '2389_CL17',\n",
       "       '2389_CL18', '2389_CL3', '2389_CL4', '2389_CL5', '2389_CL6',\n",
       "       '2389_CL7', '2389_CL9', '2542_CL1', '2542_CL10', '2542_CL11',\n",
       "       '2542_CL12', '2542_CL13', '2542_CL14', '2542_CL15', '2542_CL16',\n",
       "       '2542_CL17', '2542_CL18', '2542_CL19', '2542_CL2', '2542_CL3',\n",
       "       '2542_CL4', '2542_CL5', '2542_CL6', '2542_CL7', '2542_CL8',\n",
       "       '2542_CL9', '2819_CL1', '2819_CL10', '2819_CL11', '2819_CL12',\n",
       "       '2819_CL4', '2819_CL5', '2819_CL6', '2819_CL7', '2819_CL8',\n",
       "       '2819_CL9', '2974_CL1', '2974_CL2', '2974_CL3', '2974_CL4',\n",
       "       '2974_CL5', '2974_CL6', '2974_CL8', '2974_CL9'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_maf.sort_values(by='Tumor_Sample_Barcode')['Tumor_Sample_Barcode'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8ae345cf-23fa-4ba9-9c4a-0d3b19da062e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## sort coding maf\n",
    "coding_maf = coding_maf.sort_values(by=['Tumor_Sample_Barcode'])\n",
    "coding_maf = coding_maf.reset_index()\n",
    "coding_maf.drop(columns=['index'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d9528815-4ccf-4d52-b23c-989dfaf91cb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_Alias</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Cluster_Assignment</th>\n",
       "      <th>Allelic_CN_minor</th>\n",
       "      <th>Allelic_CN_major</th>\n",
       "      <th>preDP_ccf_mean</th>\n",
       "      <th>preDP_ccf_CI_low</th>\n",
       "      <th>preDP_ccf_CI_high</th>\n",
       "      <th>clust_ccf_mean</th>\n",
       "      <th>clust_ccf_CI_low</th>\n",
       "      <th>clust_ccf_CI_high</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1002</td>\n",
       "      <td>pre</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TLX2</td>\n",
       "      <td>2</td>\n",
       "      <td>74743128</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.59</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1002_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1002</td>\n",
       "      <td>pre</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SEC31B</td>\n",
       "      <td>10</td>\n",
       "      <td>102267765</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>90</td>\n",
       "      <td>38</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.75</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1002_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1002</td>\n",
       "      <td>pre</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OR6K6</td>\n",
       "      <td>1</td>\n",
       "      <td>158725164</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>95</td>\n",
       "      <td>38</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.73</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1002_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1002</td>\n",
       "      <td>pre</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAXC</td>\n",
       "      <td>6</td>\n",
       "      <td>99729273</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>64</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.80</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1002_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002</td>\n",
       "      <td>pre</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BAI2</td>\n",
       "      <td>1</td>\n",
       "      <td>32196555</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.28</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1002_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11099</th>\n",
       "      <td>2974</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RAVER2</td>\n",
       "      <td>1</td>\n",
       "      <td>65273093</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>285</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2974_CL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11100</th>\n",
       "      <td>2974</td>\n",
       "      <td>cfDNA_1976</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF93</td>\n",
       "      <td>19</td>\n",
       "      <td>20045112</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>113</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>2974_CL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11101</th>\n",
       "      <td>2974</td>\n",
       "      <td>cfDNA_1976</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF737</td>\n",
       "      <td>19</td>\n",
       "      <td>20728105</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>259</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>2974_CL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11102</th>\n",
       "      <td>2974</td>\n",
       "      <td>cfDNA_1976</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF729</td>\n",
       "      <td>19</td>\n",
       "      <td>22498344</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>145</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>2974_CL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11103</th>\n",
       "      <td>2974</td>\n",
       "      <td>cfDNA_1976</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTNNA2</td>\n",
       "      <td>2</td>\n",
       "      <td>80136918</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>222</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>2974_CL9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11104 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Patient_ID   Sample_ID  Sample_Alias Hugo_Symbol  Chromosome  \\\n",
       "0            1002         pre           NaN        TLX2           2   \n",
       "1            1002         pre           NaN      SEC31B          10   \n",
       "2            1002         pre           NaN       OR6K6           1   \n",
       "3            1002         pre           NaN        FAXC           6   \n",
       "4            1002         pre           NaN        BAI2           1   \n",
       "...           ...         ...           ...         ...         ...   \n",
       "11099        2974           7           NaN      RAVER2           1   \n",
       "11100        2974  cfDNA_1976           NaN       ZNF93          19   \n",
       "11101        2974  cfDNA_1976           NaN      ZNF737          19   \n",
       "11102        2974  cfDNA_1976           NaN      ZNF729          19   \n",
       "11103        2974  cfDNA_1976           NaN      CTNNA2           2   \n",
       "\n",
       "       Start_position Reference_Allele Tumor_Seq_Allele  t_ref_count  \\\n",
       "0            74743128                G                C           10   \n",
       "1           102267765                G                C           90   \n",
       "2           158725164                G                C           95   \n",
       "3            99729273                C                G           64   \n",
       "4            32196555                C                T            9   \n",
       "...               ...              ...              ...          ...   \n",
       "11099        65273093                G                A          285   \n",
       "11100        20045112                A                G          113   \n",
       "11101        20728105                C                T          259   \n",
       "11102        22498344                G                A          145   \n",
       "11103        80136918                A                C          222   \n",
       "\n",
       "       t_alt_count  ... Cluster_Assignment Allelic_CN_minor Allelic_CN_major  \\\n",
       "0                6  ...                  1              NaN              NaN   \n",
       "1               38  ...                  1              NaN              NaN   \n",
       "2               38  ...                  1              NaN              NaN   \n",
       "3               28  ...                  1              NaN              NaN   \n",
       "4                2  ...                  1              NaN              NaN   \n",
       "...            ...  ...                ...              ...              ...   \n",
       "11099            7  ...                  9              NaN              NaN   \n",
       "11100            4  ...                  9              NaN              NaN   \n",
       "11101           10  ...                  9              NaN              NaN   \n",
       "11102            3  ...                  9              NaN              NaN   \n",
       "11103            2  ...                  9              NaN              NaN   \n",
       "\n",
       "       preDP_ccf_mean  preDP_ccf_CI_low  preDP_ccf_CI_high  clust_ccf_mean  \\\n",
       "0                0.81              0.59               1.00            1.00   \n",
       "1                0.88              0.75               1.00            1.00   \n",
       "2                0.87              0.73               1.00            1.00   \n",
       "3                0.91              0.80               1.00            1.00   \n",
       "4                0.60              0.28               1.00            1.00   \n",
       "...               ...               ...                ...             ...   \n",
       "11099            0.09              0.04               0.15            0.01   \n",
       "11100            0.19              0.06               0.32            0.04   \n",
       "11101            0.18              0.08               0.28            0.04   \n",
       "11102            0.12              0.03               0.22            0.04   \n",
       "11103            0.06              0.01               0.12            0.04   \n",
       "\n",
       "       clust_ccf_CI_low  clust_ccf_CI_high  Tumor_Sample_Barcode  \n",
       "0                  0.99               1.00              1002_CL1  \n",
       "1                  0.99               1.00              1002_CL1  \n",
       "2                  0.99               1.00              1002_CL1  \n",
       "3                  0.99               1.00              1002_CL1  \n",
       "4                  0.99               1.00              1002_CL1  \n",
       "...                 ...                ...                   ...  \n",
       "11099              0.01               0.01              2974_CL9  \n",
       "11100              0.04               0.04              2974_CL9  \n",
       "11101              0.04               0.04              2974_CL9  \n",
       "11102              0.04               0.04              2974_CL9  \n",
       "11103              0.04               0.04              2974_CL9  \n",
       "\n",
       "[11104 rows x 23 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_maf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c4aa1b9-54da-4cdc-8f0e-9ebdb1feceaf",
   "metadata": {},
   "source": [
    "## test with KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7dbe4bc6-84e4-4e54-a07c-49b094117396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_Alias</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Cluster_Assignment</th>\n",
       "      <th>Allelic_CN_minor</th>\n",
       "      <th>Allelic_CN_major</th>\n",
       "      <th>preDP_ccf_mean</th>\n",
       "      <th>preDP_ccf_CI_low</th>\n",
       "      <th>preDP_ccf_CI_high</th>\n",
       "      <th>clust_ccf_mean</th>\n",
       "      <th>clust_ccf_CI_low</th>\n",
       "      <th>clust_ccf_CI_high</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4112</th>\n",
       "      <td>1326</td>\n",
       "      <td>RP-1700_BWES00001_v1_Exome_OnPrem</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151896451</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>36</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.57</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1326_CL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4145</th>\n",
       "      <td>1326</td>\n",
       "      <td>RP-1700_T01027-3_v4_Exome_OnPrem</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151945382</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>246</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1326_CL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4588</th>\n",
       "      <td>1547</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879260</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>184</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1547_CL14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4650</th>\n",
       "      <td>1547</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874772</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>180</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1547_CL15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4724</th>\n",
       "      <td>1558</td>\n",
       "      <td>RA_1558_cfDNA_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879322</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>79</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.76</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1558_CL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4780</th>\n",
       "      <td>1558</td>\n",
       "      <td>1558_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152012253</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>185</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1558_CL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5466</th>\n",
       "      <td>2389</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878626</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>122</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2389_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6202</th>\n",
       "      <td>2389</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152007146</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6379</th>\n",
       "      <td>2389</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878044</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>194</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6487</th>\n",
       "      <td>2389</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874215</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6581</th>\n",
       "      <td>2389</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151917668</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>189</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6682</th>\n",
       "      <td>2389</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151859429</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>185</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6755</th>\n",
       "      <td>2389</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151880082</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>88</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2389_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7069</th>\n",
       "      <td>2389</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151851222</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>158</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.17</td>\n",
       "      <td>2389_CL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7115</th>\n",
       "      <td>2389</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151891593</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>101</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.90</td>\n",
       "      <td>2389_CL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7208</th>\n",
       "      <td>2389</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151874161</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>152</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2389_CL7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7925</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151871322</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>133</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.64</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>2542_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8053</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151878785</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>185</td>\n",
       "      <td>42</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.75</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>2542_CL12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9089</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>152007108</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>234</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.23</td>\n",
       "      <td>2542_CL15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10047</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151879670</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.48</td>\n",
       "      <td>2542_CL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10410</th>\n",
       "      <td>2542</td>\n",
       "      <td>2542_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151932951</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>357</td>\n",
       "      <td>53</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.63</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.98</td>\n",
       "      <td>2542_CL2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Patient_ID                          Sample_ID  Sample_Alias  \\\n",
       "4112         1326  RP-1700_BWES00001_v1_Exome_OnPrem           NaN   \n",
       "4145         1326   RP-1700_T01027-3_v4_Exome_OnPrem           NaN   \n",
       "4588         1547                                 10           NaN   \n",
       "4650         1547                                  6           NaN   \n",
       "4724         1558                    RA_1558_cfDNA_3           NaN   \n",
       "4780         1558                             1558_6           NaN   \n",
       "5466         2389                                  4           NaN   \n",
       "6202         2389                                  8           NaN   \n",
       "6379         2389                                  7           NaN   \n",
       "6487         2389                                 11           NaN   \n",
       "6581         2389                                 10           NaN   \n",
       "6682         2389                                  3           NaN   \n",
       "6755         2389                                 18           NaN   \n",
       "7069         2389                                  4           NaN   \n",
       "7115         2389                                  1           NaN   \n",
       "7208         2389                                  8           NaN   \n",
       "7925         2542                             2542_1           NaN   \n",
       "8053         2542                             2542_1           NaN   \n",
       "9089         2542                             2542_8           NaN   \n",
       "10047        2542                             2542_2           NaN   \n",
       "10410        2542                             2542_2           NaN   \n",
       "\n",
       "      Hugo_Symbol  Chromosome  Start_position Reference_Allele  \\\n",
       "4112        KMT2C           7       151896451                T   \n",
       "4145        KMT2C           7       151945382                C   \n",
       "4588        KMT2C           7       151879260                C   \n",
       "4650        KMT2C           7       151874772                C   \n",
       "4724        KMT2C           7       151879322                G   \n",
       "4780        KMT2C           7       152012253                G   \n",
       "5466        KMT2C           7       151878626                C   \n",
       "6202        KMT2C           7       152007146                G   \n",
       "6379        KMT2C           7       151878044                G   \n",
       "6487        KMT2C           7       151874215                C   \n",
       "6581        KMT2C           7       151917668                G   \n",
       "6682        KMT2C           7       151859429                C   \n",
       "6755        KMT2C           7       151880082                C   \n",
       "7069        KMT2C           7       151851222                G   \n",
       "7115        KMT2C           7       151891593                G   \n",
       "7208        KMT2C           7       151874161                C   \n",
       "7925        KMT2C           7       151871322                C   \n",
       "8053        KMT2C           7       151878785                G   \n",
       "9089        KMT2C           7       152007108                C   \n",
       "10047       KMT2C           7       151879670                G   \n",
       "10410       KMT2C           7       151932951                C   \n",
       "\n",
       "      Tumor_Seq_Allele  t_ref_count  t_alt_count  ... Cluster_Assignment  \\\n",
       "4112                 A           36           13  ...                  2   \n",
       "4145                 -          246           64  ...                  9   \n",
       "4588                 A          184           12  ...                 14   \n",
       "4650                 T          180            6  ...                 15   \n",
       "4724                 A           79           33  ...                  1   \n",
       "4780                 C          185           21  ...                 10   \n",
       "5466                 A          122            5  ...                 12   \n",
       "6202                 A          211            1  ...                 17   \n",
       "6379                 A          194            1  ...                 18   \n",
       "6487                 A           61            1  ...                 18   \n",
       "6581                 C          189            1  ...                 18   \n",
       "6682                 A          185            4  ...                 18   \n",
       "6755                 A           88            7  ...                 18   \n",
       "7069                 C          158           10  ...                  4   \n",
       "7115                 A          101           46  ...                  5   \n",
       "7208                 A          152            1  ...                  7   \n",
       "7925                 A          133           25  ...                 12   \n",
       "8053                 A          185           42  ...                 12   \n",
       "9089                 G          234            6  ...                 15   \n",
       "10047                A           30            3  ...                 18   \n",
       "10410                T          357           53  ...                  2   \n",
       "\n",
       "      Allelic_CN_minor Allelic_CN_major  preDP_ccf_mean  preDP_ccf_CI_low  \\\n",
       "4112               NaN              NaN            0.80              0.57   \n",
       "4145               NaN              NaN            0.87              0.74   \n",
       "4588               NaN              NaN            0.18              0.10   \n",
       "4650               NaN              NaN            0.09              0.04   \n",
       "4724               NaN              NaN            0.89              0.76   \n",
       "4780               NaN              NaN            0.39              0.25   \n",
       "5466               NaN              NaN            0.16              0.06   \n",
       "6202               NaN              NaN            0.34              0.01   \n",
       "6379               NaN              NaN            0.05              0.01   \n",
       "6487               NaN              NaN            0.06              0.01   \n",
       "6581               NaN              NaN            0.03              0.01   \n",
       "6682               NaN              NaN            0.08              0.03   \n",
       "6755               NaN              NaN            0.47              0.19   \n",
       "7069               NaN              NaN            0.22              0.11   \n",
       "7115               NaN              NaN            0.77              0.62   \n",
       "7208               NaN              NaN            0.38              0.01   \n",
       "7925               NaN              NaN            0.83              0.64   \n",
       "8053               NaN              NaN            0.88              0.75   \n",
       "9089               NaN              NaN            0.39              0.13   \n",
       "10047              NaN              NaN            0.56              0.24   \n",
       "10410              NaN              NaN            0.81              0.63   \n",
       "\n",
       "       preDP_ccf_CI_high  clust_ccf_mean  clust_ccf_CI_low  clust_ccf_CI_high  \\\n",
       "4112                1.00            0.98              0.96               1.00   \n",
       "4145                1.00            0.83              0.77               0.89   \n",
       "4588                0.27            0.06              0.06               0.06   \n",
       "4650                0.14            0.05              0.05               0.06   \n",
       "4724                1.00            1.00              0.99               1.00   \n",
       "4780                0.53            0.30              0.29               0.30   \n",
       "5466                0.25            0.36              0.35               0.37   \n",
       "6202                0.78            0.01              0.01               0.01   \n",
       "6379                0.09            0.01              0.01               0.01   \n",
       "6487                0.10            0.01              0.01               0.01   \n",
       "6581                0.05            0.01              0.01               0.01   \n",
       "6682                0.12            0.01              0.01               0.01   \n",
       "6755                0.74            0.03              0.03               0.03   \n",
       "7069                0.36            0.16              0.15               0.17   \n",
       "7115                0.92            0.88              0.86               0.90   \n",
       "7208                0.82            0.03              0.02               0.03   \n",
       "7925                1.00            0.86              0.86               0.86   \n",
       "8053                1.00            0.86              0.86               0.86   \n",
       "9089                0.64            0.23              0.22               0.23   \n",
       "10047               0.95            0.48              0.47               0.48   \n",
       "10410               1.00            0.94              0.91               0.98   \n",
       "\n",
       "       Tumor_Sample_Barcode  \n",
       "4112               1326_CL2  \n",
       "4145               1326_CL9  \n",
       "4588              1547_CL14  \n",
       "4650              1547_CL15  \n",
       "4724               1558_CL1  \n",
       "4780              1558_CL10  \n",
       "5466              2389_CL12  \n",
       "6202              2389_CL17  \n",
       "6379              2389_CL18  \n",
       "6487              2389_CL18  \n",
       "6581              2389_CL18  \n",
       "6682              2389_CL18  \n",
       "6755              2389_CL18  \n",
       "7069               2389_CL4  \n",
       "7115               2389_CL5  \n",
       "7208               2389_CL7  \n",
       "7925              2542_CL12  \n",
       "8053              2542_CL12  \n",
       "9089              2542_CL15  \n",
       "10047             2542_CL18  \n",
       "10410              2542_CL2  \n",
       "\n",
       "[21 rows x 23 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KMT2C_all_muts = coding_maf[coding_maf['Hugo_Symbol']=='KMT2C']\n",
    "KMT2C_all_muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "40eb2b3d-602a-4ee8-ab49-abb4f666ac7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## total number of mutations in the target gene N0\n",
    "N0 = len(KMT2C_all_muts)\n",
    "## total number of clones\n",
    "N_c = len( coding_maf['Tumor_Sample_Barcode'].unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bb207faa-5e08-4ec0-8735-1db2f8c678b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0f17c32-ceed-4478-aa27-ecdb67d9d990",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "159"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N_c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc1fb54-96e6-4f61-90ec-965a4821412c",
   "metadata": {},
   "source": [
    "## start with one patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "52bed5fa-6e1c-4bf7-abc3-4aac4962d2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "RA_2389 = coding_maf[coding_maf['Patient_ID']==2389]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af3465be-5842-4af8-b541-bd1ceb4d3cd5",
   "metadata": {},
   "source": [
    "### for every clone, I am going to model the probability that KMT2C is in the clone or not\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "dea32333-95a7-40ef-9f51-da2d0668dce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "RA_2389_clones = list(RA_2389['Tumor_Sample_Barcode'].unique() )\n",
    "# total number of mutations in this patient\n",
    "M = len(RA_2389)\n",
    " \n",
    "# initialize a Pij matrix where you have a 1 if a clone was mutated and then the score\n",
    "P_2389 =np.zeros( (len(RA_2389_clones), 2 ))\n",
    "\n",
    "gene=\"KMT2C\"\n",
    "\n",
    "clone_idx=0\n",
    "for clone in RA_2389_clones:\n",
    "    clone_df = RA_2389[RA_2389['Tumor_Sample_Barcode'] == clone]\n",
    "    mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "    Mij = len(list(mutations_in_clone))\n",
    "    \n",
    "    term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "    \n",
    "    # a given clone has no mutation in the gene\n",
    "    P_2389[clone_idx, 0] =  term_to_calc\n",
    "    \n",
    "    # a given clone has one or more mutations in the gene\n",
    "    P_2389[clone_idx, 1] = 1- term_to_calc\n",
    "    \n",
    "\n",
    "    clone_idx +=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9241ecb8-4e10-4904-b77c-7e70602fe795",
   "metadata": {},
   "outputs": [],
   "source": [
    "## now every row sums to one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2ed36da2-9c77-42f9-ab5e-c066a65d19d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>no_KMT2C</th>\n",
       "      <th>KMT2C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2389_CL16</td>\n",
       "      <td>0.862730</td>\n",
       "      <td>0.137270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2389_CL5</td>\n",
       "      <td>0.938615</td>\n",
       "      <td>0.061385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2389_CL3</td>\n",
       "      <td>0.927032</td>\n",
       "      <td>0.072968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2389_CL18</td>\n",
       "      <td>0.359055</td>\n",
       "      <td>0.640945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2389_CL14</td>\n",
       "      <td>0.916851</td>\n",
       "      <td>0.083149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2389_CL15</td>\n",
       "      <td>0.758178</td>\n",
       "      <td>0.241822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2389_CL1</td>\n",
       "      <td>0.979575</td>\n",
       "      <td>0.020425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2389_CL17</td>\n",
       "      <td>0.834506</td>\n",
       "      <td>0.165494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2389_CL6</td>\n",
       "      <td>0.920656</td>\n",
       "      <td>0.079344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2389_CL7</td>\n",
       "      <td>0.939910</td>\n",
       "      <td>0.060090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2389_CL4</td>\n",
       "      <td>0.967511</td>\n",
       "      <td>0.032489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2389_CL9</td>\n",
       "      <td>0.877180</td>\n",
       "      <td>0.122820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2389_CL13</td>\n",
       "      <td>0.759235</td>\n",
       "      <td>0.240765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2389_CL12</td>\n",
       "      <td>0.891861</td>\n",
       "      <td>0.108139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2389_CL10</td>\n",
       "      <td>0.970180</td>\n",
       "      <td>0.029820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2389_CL11</td>\n",
       "      <td>0.914322</td>\n",
       "      <td>0.085678</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        clone  no_KMT2C     KMT2C\n",
       "0   2389_CL16  0.862730  0.137270\n",
       "1    2389_CL5  0.938615  0.061385\n",
       "2    2389_CL3  0.927032  0.072968\n",
       "3   2389_CL18  0.359055  0.640945\n",
       "4   2389_CL14  0.916851  0.083149\n",
       "5   2389_CL15  0.758178  0.241822\n",
       "6    2389_CL1  0.979575  0.020425\n",
       "7   2389_CL17  0.834506  0.165494\n",
       "8    2389_CL6  0.920656  0.079344\n",
       "9    2389_CL7  0.939910  0.060090\n",
       "10   2389_CL4  0.967511  0.032489\n",
       "11   2389_CL9  0.877180  0.122820\n",
       "12  2389_CL13  0.759235  0.240765\n",
       "13  2389_CL12  0.891861  0.108139\n",
       "14  2389_CL10  0.970180  0.029820\n",
       "15  2389_CL11  0.914322  0.085678"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_2389_prob_df = pd.DataFrame(P_2389, columns=['no_KMT2C', 'KMT2C'])\n",
    "P_2389_prob_df.insert(0, 'clone', RA_2389_clones)\n",
    "P_2389_prob_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "3f8ade90-6e5f-408e-aaaa-1a90c5303e40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_Alias</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Cluster_Assignment</th>\n",
       "      <th>Allelic_CN_minor</th>\n",
       "      <th>Allelic_CN_major</th>\n",
       "      <th>preDP_ccf_mean</th>\n",
       "      <th>preDP_ccf_CI_low</th>\n",
       "      <th>preDP_ccf_CI_high</th>\n",
       "      <th>clust_ccf_mean</th>\n",
       "      <th>clust_ccf_CI_low</th>\n",
       "      <th>clust_ccf_CI_high</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2389_CL1</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL10</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL11</th>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>...</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL12</th>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>...</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL13</th>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>0</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>...</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL14</th>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL15</th>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>0</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>...</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL16</th>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>...</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL17</th>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>...</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL18</th>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>0</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>...</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "      <td>775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL3</th>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>...</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL4</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL5</th>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL6</th>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL7</th>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389_CL9</th>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>...</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Patient_ID  Sample_ID  Sample_Alias  Hugo_Symbol  \\\n",
       "Tumor_Sample_Barcode                                                     \n",
       "2389_CL1                      17         17             0           17   \n",
       "2389_CL10                     22         22             0           22   \n",
       "2389_CL11                     65         65             0           65   \n",
       "2389_CL12                     87         87             0           87   \n",
       "2389_CL13                    206        206             0          206   \n",
       "2389_CL14                     64         64             0           64   \n",
       "2389_CL15                    205        205             0          205   \n",
       "2389_CL16                    112        112             0          112   \n",
       "2389_CL17                    133        133             0          133   \n",
       "2389_CL18                    775        775             0          775   \n",
       "2389_CL3                      57         57             0           57   \n",
       "2389_CL4                      24         24             0           24   \n",
       "2389_CL5                      46         46             0           46   \n",
       "2389_CL6                      60         60             0           60   \n",
       "2389_CL7                      46         46             0           46   \n",
       "2389_CL9                      98         98             0           98   \n",
       "\n",
       "                      Chromosome  Start_position  Reference_Allele  \\\n",
       "Tumor_Sample_Barcode                                                 \n",
       "2389_CL1                      17              17                17   \n",
       "2389_CL10                     22              22                22   \n",
       "2389_CL11                     65              65                65   \n",
       "2389_CL12                     87              87                87   \n",
       "2389_CL13                    206             206               206   \n",
       "2389_CL14                     64              64                64   \n",
       "2389_CL15                    205             205               205   \n",
       "2389_CL16                    112             112               112   \n",
       "2389_CL17                    133             133               133   \n",
       "2389_CL18                    775             775               775   \n",
       "2389_CL3                      57              57                57   \n",
       "2389_CL4                      24              24                24   \n",
       "2389_CL5                      46              46                46   \n",
       "2389_CL6                      60              60                60   \n",
       "2389_CL7                      46              46                46   \n",
       "2389_CL9                      98              98                98   \n",
       "\n",
       "                      Tumor_Seq_Allele  t_ref_count  t_alt_count  ...  \\\n",
       "Tumor_Sample_Barcode                                              ...   \n",
       "2389_CL1                            17           17           17  ...   \n",
       "2389_CL10                           22           22           22  ...   \n",
       "2389_CL11                           65           65           65  ...   \n",
       "2389_CL12                           87           87           87  ...   \n",
       "2389_CL13                          206          206          206  ...   \n",
       "2389_CL14                           64           64           64  ...   \n",
       "2389_CL15                          205          205          205  ...   \n",
       "2389_CL16                          112          112          112  ...   \n",
       "2389_CL17                          133          133          133  ...   \n",
       "2389_CL18                          775          775          775  ...   \n",
       "2389_CL3                            57           57           57  ...   \n",
       "2389_CL4                            24           24           24  ...   \n",
       "2389_CL5                            46           46           46  ...   \n",
       "2389_CL6                            60           60           60  ...   \n",
       "2389_CL7                            46           46           46  ...   \n",
       "2389_CL9                            98           98           98  ...   \n",
       "\n",
       "                      Variant_Type  Cluster_Assignment  Allelic_CN_minor  \\\n",
       "Tumor_Sample_Barcode                                                       \n",
       "2389_CL1                        17                  17                 0   \n",
       "2389_CL10                       22                  22                 0   \n",
       "2389_CL11                       65                  65                 0   \n",
       "2389_CL12                       87                  87                 0   \n",
       "2389_CL13                      206                 206                 0   \n",
       "2389_CL14                       64                  64                 0   \n",
       "2389_CL15                      205                 205                 0   \n",
       "2389_CL16                      112                 112                 0   \n",
       "2389_CL17                      133                 133                 0   \n",
       "2389_CL18                      775                 775                 0   \n",
       "2389_CL3                        57                  57                 0   \n",
       "2389_CL4                        24                  24                 0   \n",
       "2389_CL5                        46                  46                 0   \n",
       "2389_CL6                        60                  60                 0   \n",
       "2389_CL7                        46                  46                 0   \n",
       "2389_CL9                        98                  98                 0   \n",
       "\n",
       "                      Allelic_CN_major  preDP_ccf_mean  preDP_ccf_CI_low  \\\n",
       "Tumor_Sample_Barcode                                                       \n",
       "2389_CL1                             0              17                17   \n",
       "2389_CL10                            0              22                22   \n",
       "2389_CL11                            0              65                65   \n",
       "2389_CL12                            0              87                87   \n",
       "2389_CL13                            0             206               206   \n",
       "2389_CL14                            0              64                64   \n",
       "2389_CL15                            0             205               205   \n",
       "2389_CL16                            0             112               112   \n",
       "2389_CL17                            0             133               133   \n",
       "2389_CL18                            0             775               775   \n",
       "2389_CL3                             0              57                57   \n",
       "2389_CL4                             0              24                24   \n",
       "2389_CL5                             0              46                46   \n",
       "2389_CL6                             0              60                60   \n",
       "2389_CL7                             0              46                46   \n",
       "2389_CL9                             0              98                98   \n",
       "\n",
       "                      preDP_ccf_CI_high  clust_ccf_mean  clust_ccf_CI_low  \\\n",
       "Tumor_Sample_Barcode                                                        \n",
       "2389_CL1                             17              17                17   \n",
       "2389_CL10                            22              22                22   \n",
       "2389_CL11                            65              65                65   \n",
       "2389_CL12                            87              87                87   \n",
       "2389_CL13                           206             206               206   \n",
       "2389_CL14                            64              64                64   \n",
       "2389_CL15                           205             205               205   \n",
       "2389_CL16                           112             112               112   \n",
       "2389_CL17                           133             133               133   \n",
       "2389_CL18                           775             775               775   \n",
       "2389_CL3                             57              57                57   \n",
       "2389_CL4                             24              24                24   \n",
       "2389_CL5                             46              46                46   \n",
       "2389_CL6                             60              60                60   \n",
       "2389_CL7                             46              46                46   \n",
       "2389_CL9                             98              98                98   \n",
       "\n",
       "                      clust_ccf_CI_high  \n",
       "Tumor_Sample_Barcode                     \n",
       "2389_CL1                             17  \n",
       "2389_CL10                            22  \n",
       "2389_CL11                            65  \n",
       "2389_CL12                            87  \n",
       "2389_CL13                           206  \n",
       "2389_CL14                            64  \n",
       "2389_CL15                           205  \n",
       "2389_CL16                           112  \n",
       "2389_CL17                           133  \n",
       "2389_CL18                           775  \n",
       "2389_CL3                             57  \n",
       "2389_CL4                             24  \n",
       "2389_CL5                             46  \n",
       "2389_CL6                             60  \n",
       "2389_CL7                             46  \n",
       "2389_CL9                             98  \n",
       "\n",
       "[16 rows x 22 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RA_2389.groupby('Tumor_Sample_Barcode').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4caa1ec-5c59-449c-b283-a345e20a6959",
   "metadata": {},
   "source": [
    "## patient score if no KMT2C mutation in any clone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6af28654-348a-4d47-b470-13b92097a7a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.6946982684320844\n"
     ]
    }
   ],
   "source": [
    "P_2389_prob_df_no_KMT2C = P_2389_prob_df['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C = [ np.log(p) for p in P_2389_prob_df_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8fab121-cf22-4aa9-8d20-1ff5a407d6af",
   "metadata": {},
   "source": [
    "## patient score if a KMT2C mutation in all clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e3252520-4456-4413-9daa-4eb7ed82fae5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37.92316367280446\n"
     ]
    }
   ],
   "source": [
    "P_2389_prob_df_KMT2C = P_2389_prob_df['KMT2C'].to_list()\n",
    "logs_KMT2C = [ np.log(p) for p in P_2389_prob_df_KMT2C]\n",
    "print( -np.sum(logs_KMT2C))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32568f76-ad33-4f8a-aecf-758a986c41e3",
   "metadata": {},
   "source": [
    "## test RA 1598, no KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ff0149b2-f0c5-4ab7-adb7-1e4c9f52f7e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "RA_1598= coding_maf[coding_maf['Patient_ID']==1598]\n",
    "RA_1598_clones = list(RA_1598['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "# total number of mutations in this patient\n",
    "M = len(RA_1598)\n",
    "# initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "P_1598 =np.zeros( (len(RA_1598_clones), 2 ))\n",
    "\n",
    "gene=\"KMT2C\"\n",
    "\n",
    "clone_idx=0\n",
    "for clone in RA_1598_clones:\n",
    "    clone_df = RA_1598[RA_1598['Tumor_Sample_Barcode'] == clone]\n",
    "    mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "    Mij = len(list(mutations_in_clone))\n",
    "    term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "    # a given clone has no mutation in the gene\n",
    "    P_1598[clone_idx, 0] =  term_to_calc\n",
    "    \n",
    "    #  a given clone has a mutation in the gene\n",
    "    P_1598[clone_idx, 1] = 1-term_to_calc\n",
    "    \n",
    "\n",
    "    clone_idx +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b8233f8a-9598-4d47-a4a0-6f3d4c550d79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>no_KMT2C</th>\n",
       "      <th>KMT2C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1598_CL2</td>\n",
       "      <td>0.931663</td>\n",
       "      <td>0.068337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1598_CL1</td>\n",
       "      <td>0.867789</td>\n",
       "      <td>0.132211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1598_CL8</td>\n",
       "      <td>0.700242</td>\n",
       "      <td>0.299758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1598_CL10</td>\n",
       "      <td>0.269250</td>\n",
       "      <td>0.730750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1598_CL5</td>\n",
       "      <td>0.883355</td>\n",
       "      <td>0.116645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1598_CL4</td>\n",
       "      <td>0.931663</td>\n",
       "      <td>0.068337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1598_CL7</td>\n",
       "      <td>0.793799</td>\n",
       "      <td>0.206201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1598_CL9</td>\n",
       "      <td>0.867789</td>\n",
       "      <td>0.132211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1598_CL3</td>\n",
       "      <td>0.899186</td>\n",
       "      <td>0.100814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1598_CL6</td>\n",
       "      <td>0.793799</td>\n",
       "      <td>0.206201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       clone  no_KMT2C     KMT2C\n",
       "0   1598_CL2  0.931663  0.068337\n",
       "1   1598_CL1  0.867789  0.132211\n",
       "2   1598_CL8  0.700242  0.299758\n",
       "3  1598_CL10  0.269250  0.730750\n",
       "4   1598_CL5  0.883355  0.116645\n",
       "5   1598_CL4  0.931663  0.068337\n",
       "6   1598_CL7  0.793799  0.206201\n",
       "7   1598_CL9  0.867789  0.132211\n",
       "8   1598_CL3  0.899186  0.100814\n",
       "9   1598_CL6  0.793799  0.206201"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_1598_prob_df = pd.DataFrame(P_1598, columns=['no_KMT2C', 'KMT2C'])\n",
    "P_1598_prob_df.insert(0, 'clone', RA_1598_clones)\n",
    "P_1598_prob_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16242eda-896b-4dc8-aee1-1f438e56f3a8",
   "metadata": {},
   "source": [
    "## patient score if no KMT2C clones (truth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "abb3dc3b-4b06-4265-9d51-a92e05ee103d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.785767361195695\n"
     ]
    }
   ],
   "source": [
    "P_1598_prob_df_no_KMT2C = P_1598_prob_df['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C_1598 = [ np.log(p) for p in P_1598_prob_df_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C_1598))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ccf5843-6c34-4980-a4c8-cfb48e4a1b0c",
   "metadata": {},
   "source": [
    "## patient score if all KMT2C clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3465bf8e-6f37-4fb6-9d6f-fa7bdd57feef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18.532696469275578\n"
     ]
    }
   ],
   "source": [
    "P_1598_prob_df_KMT2C = P_1598_prob_df['KMT2C'].to_list()\n",
    "logs_KMT2C_1598 = [ np.log(p) for p in P_1598_prob_df_KMT2C]\n",
    "print( -np.sum(logs_KMT2C_1598))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7e110c6-a2d7-4e39-97ee-2aa317bd4c88",
   "metadata": {},
   "source": [
    "## try for all patients for KMT2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "id": "25f06c03-9d72-40cc-815c-b4816af8e0a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "id": "4c0f1cb5-7e75-4c54-88f3-17209c400de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_probs_KMT2C = []\n",
    "\n",
    "for patient in all_patients: \n",
    "\n",
    "    RA_maf = coding_maf[coding_maf['Patient_ID']==patient]\n",
    "    RA_maf_clones = list(RA_maf['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "    ## total number of mutations in the target gene N0\n",
    "    N0 = len(KMT2C_all_muts)\n",
    "    ## total number of clones\n",
    "    N_c = len( coding_maf['Tumor_Sample_Barcode'].unique() )\n",
    "\n",
    "    # total number of mutations in this patient\n",
    "    M = len(RA_maf)\n",
    "    # initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "    P_pat =np.zeros( (len(RA_maf_clones), 2 ))\n",
    "    \n",
    "    gene=\"KMT2C\"\n",
    "    \n",
    "    clone_idx=0\n",
    "    for clone in RA_maf_clones:\n",
    "        clone_df = RA_maf[RA_maf['Tumor_Sample_Barcode'] == clone]\n",
    "        mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "        Mij = len(list(mutations_in_clone))\n",
    "        \n",
    "        term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "        \n",
    "        # probability that a given clone has no mutation in the gene\n",
    "        P_pat[clone_idx, 0] = term_to_calc\n",
    "        \n",
    "        # probability that a given clone has one or more mutations in the gene\n",
    "        P_pat[clone_idx, 1] = 1- term_to_calc\n",
    "        clone_idx +=1\n",
    "\n",
    "    P_prob_df = pd.DataFrame(P_pat, columns=['no_KMT2C', 'KMT2C'])\n",
    "    P_prob_df.insert(0, 'clone', RA_maf_clones)\n",
    "    all_probs_KMT2C.append(P_prob_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8ac75a0-00cf-464a-befe-8f6dbf6d9022",
   "metadata": {},
   "source": [
    "## score if no KMT2C mutation in this patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6b9ae65e-4bad-4a09-81c3-f66e005cd019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.785239956066282\n"
     ]
    }
   ],
   "source": [
    "probs_1326_no_KMT2C = probs_1326['no_KMT2C'].to_list()\n",
    "logs_no_KMT2C_1326 = [ np.log(p) for p in probs_1326_no_KMT2C]\n",
    "print( -np.sum(logs_no_KMT2C_1326))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a69120-e34b-472d-bc14-4a6bcbb142d9",
   "metadata": {},
   "source": [
    "## real distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4ce52f65-5230-4dd1-8915-f1fb2020a9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "KMT2C_clones_RA1326= ['1326_CL2','1326_CL9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "39236529-1bb1-481e-b831-d50c9231b90c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.547794414031647"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_1326_prob_df_KMT2C_clones = probs_1326[probs_1326['clone'].isin(KMT2C_clones_RA1326)]\n",
    "P_1326_prob_df_non_KMT2C_clones = probs_1326[~probs_1326['clone'].isin(KMT2C_clones_RA1326)]\n",
    "\n",
    "P_1326_prob_df_KMT2C_clones_list = P_1326_prob_df_KMT2C_clones['KMT2C'].to_list()\n",
    "logs_real_KMT2C = [ -np.log(p) for p in P_1326_prob_df_KMT2C_clones_list]\n",
    "\n",
    "P_1326_no_KMT2C_clones_list = P_1326_prob_df_non_KMT2C_clones['no_KMT2C'].to_list()\n",
    "logs_real_no_KMT2C = [ -np.log(p) for p in P_1326_no_KMT2C_clones_list]\n",
    "\n",
    "\n",
    "np.sum( np.concatenate((logs_real_KMT2C, logs_real_no_KMT2C)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "30603c03-8872-4ba5-83f8-b52293d4581c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pats_with_no_KMT2C = [1598, 1078, 1045, 1113, 1644, 2819, 2974, 1035,1534,1002]\n",
    "pats_with_no_KMT2C_idx =[1, 2, 3, 4, 5, 6, 7, 8, 9, 14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b8956df6-9b6b-412a-b0a0-e71d74e31b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs_pats_no_KMT2C = []\n",
    "probs_pats_no_KMT2C_scores = []\n",
    "probs_pats_no_KMT2C_scores_dict = {}\n",
    "for idx in pats_with_no_KMT2C_idx:\n",
    "    score = 0\n",
    "    probs_pats_no_KMT2C.append(all_probs_KMT2C[idx])\n",
    "    pat_probs = all_probs_KMT2C[idx]\n",
    "    pat_probs_no_KMT2C = pat_probs['no_KMT2C'].to_list()\n",
    "    logs_no_KMT2C = [ np.log(p) for p in pat_probs_no_KMT2C]\n",
    "    score = -np.sum(logs_no_KMT2C)\n",
    "    \n",
    "    probs_pats_no_KMT2C_scores.append(score)\n",
    "    pat = all_probs_order[idx]\n",
    "    probs_pats_no_KMT2C_scores_dict[pat] = score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a0ed914d-ec21-466d-bd21-e57ed1156d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "pats_with_KMT2C_idx=[0, 10,11,12,13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "708331f5-7dad-4ee2-bbc3-13cad03247ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_KMT2C_clones = list(KMT2C_all_muts['Tumor_Sample_Barcode'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ab44785a-578a-4620-9321-7b8197f33edc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "probs_pats_with_KMT2C = []\n",
    "probs_pats_KMT2C_scores = []\n",
    "probs_pats_with_KMT2C_scores_dict = {}\n",
    "for idx in pats_with_KMT2C_idx:\n",
    "    #score=0\n",
    "    running_score=[]\n",
    "    pat_probs = all_probs_KMT2C[idx]\n",
    "    all_clones_in_pat = list(pat_probs['clone'])\n",
    "    for clone in all_clones_in_pat:\n",
    "        pat_probs_clone = pat_probs[pat_probs['clone']==clone]\n",
    "        if clone in all_KMT2C_clones:\n",
    "\n",
    "            pat_probs_clone_KMT2C = pat_probs_clone['KMT2C'].to_list()\n",
    "            logs_KMT2C = [ - np.log(p) for p in pat_probs_clone_KMT2C]\n",
    "            running_score.extend(logs_KMT2C)\n",
    "        else:\n",
    "            pat_probs_no_KMT2C = pat_probs_clone['no_KMT2C'].to_list()\n",
    "            logs_no_KMT2C = [ - np.log(p) for p in pat_probs_no_KMT2C]\n",
    "            running_score.extend(logs_no_KMT2C)\n",
    "            \n",
    "    score = np.sum(running_score)\n",
    "    \n",
    "    probs_pats_with_KMT2C.append(pat_probs)\n",
    "    probs_pats_KMT2C_scores.append(score)\n",
    "    pat = all_probs_order[idx]\n",
    "    probs_pats_with_KMT2C_scores_dict[pat] = score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "458ad730-0f9e-4ba0-bd9c-fcedca07be6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1547: 2.8031613827454733,\n",
       " 1558: 3.403834833843852,\n",
       " 1326: 4.547794414031647,\n",
       " 2389: 14.714030365894862,\n",
       " 2542: 8.820179860729803}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs_pats_with_KMT2C_scores_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f33186-5ad8-4705-8964-715c39980a16",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41d08f18-d731-4224-b6a3-039f3e617eb3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18cc4439-53d7-48f2-87fa-7053a21739b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0af8a38a-bab1-4852-885b-23145742a166",
   "metadata": {},
   "source": [
    "## test any gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9969e7cb-1f17-49dd-ad95-741ac1b55640",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_distributions(gene):\n",
    "    all_patients = list(coding_maf['Patient_ID'].unique())\n",
    "    gene_all_muts = coding_maf[coding_maf['Hugo_Symbol']==gene]\n",
    "    all_probs = []\n",
    "    \n",
    "    for patient in all_patients: \n",
    "    \n",
    "        RA_maf = coding_maf[coding_maf['Patient_ID']==patient]\n",
    "        RA_maf_clones = list(RA_maf['Tumor_Sample_Barcode'].unique() )\n",
    "        \n",
    "        ## total number of mutations N0\n",
    "        N0 = len(gene_all_muts)\n",
    "        ## total number of mutated clones N_c\n",
    "        N_c = len( coding_maf['Tumor_Sample_Barcode'].unique() )\n",
    "        \n",
    "        # total number of mutations in this patient\n",
    "        M = len(RA_maf)\n",
    "        # initialize a Pij matrix where you have a 1 if a clone was mutated \n",
    "        P_pat =np.zeros( (len(RA_maf_clones), 2 ))\n",
    "        \n",
    "        clone_idx=0\n",
    "        for clone in RA_maf_clones:\n",
    "            clone_df = RA_maf[RA_maf['Tumor_Sample_Barcode'] == clone]\n",
    "            mutations_in_clone = clone_df['Hugo_Symbol'].unique()\n",
    "            Mij = len(list(mutations_in_clone))\n",
    "            term_to_calc = (1 - ((Mij/M) * (N0/N_c)) ) ** N0\n",
    "\n",
    "            # probability that a given clone has no mutation in the gene\n",
    "            P_pat[clone_idx, 0] = term_to_calc\n",
    "            \n",
    "            \n",
    "            # probability that a given clone has a mutation in the gene\n",
    "            P_pat[clone_idx, 1] = 1-term_to_calc\n",
    "            clone_idx +=1\n",
    "\n",
    "    \n",
    "        P_prob_df = pd.DataFrame(P_pat, columns=[0, 1])\n",
    "        P_prob_df.insert(0, 'clone', RA_maf_clones)\n",
    "        all_probs.append(P_prob_df)\n",
    "    \n",
    "    all_probs_concat = pd.concat(all_probs)\n",
    "    pats_label=[]\n",
    "    clone_labels = list(all_probs_concat['clone'])\n",
    "    for clone in clone_labels:\n",
    "        pat = clone.split('_')[0]\n",
    "        pats_label.append(pat)\n",
    "\n",
    "    all_probs_df_labeled=all_probs_concat.copy()\n",
    "    all_probs_df_labeled['patient']=pats_label\n",
    "    all_probs_df_labeled=all_probs_df_labeled.sort_values(by=['patient'])\n",
    "    all_probs_df_labeled.reset_index(inplace=True)\n",
    "    all_probs_df_labeled.drop(columns=['index'],inplace=True)\n",
    "    return all_probs_df_labeled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "65f550ed-4630-47af-9f0f-900274af9576",
   "metadata": {},
   "outputs": [],
   "source": [
    "esr1_Pm = calculate_distributions('ESR1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5a3ddd61-6b24-4f95-a624-b6f6f1f2f2b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmt2c_Pm = calculate_distributions('KMT2C')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "95b8beac-0ea2-4568-b94f-c8ecd5ad38b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>patient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1002_CL1</td>\n",
       "      <td>0.961869</td>\n",
       "      <td>0.038131</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1002_CL6</td>\n",
       "      <td>0.995255</td>\n",
       "      <td>0.004745</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1002_CL5</td>\n",
       "      <td>0.992890</td>\n",
       "      <td>0.007110</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1002_CL3</td>\n",
       "      <td>0.987393</td>\n",
       "      <td>0.012607</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002_CL21</td>\n",
       "      <td>0.203368</td>\n",
       "      <td>0.796632</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>2974_CL3</td>\n",
       "      <td>0.955739</td>\n",
       "      <td>0.044261</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>2974_CL4</td>\n",
       "      <td>0.913348</td>\n",
       "      <td>0.086652</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>2974_CL5</td>\n",
       "      <td>0.833877</td>\n",
       "      <td>0.166123</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>2974_CL6</td>\n",
       "      <td>0.955739</td>\n",
       "      <td>0.044261</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>2974_CL9</td>\n",
       "      <td>0.509692</td>\n",
       "      <td>0.490308</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>159 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         clone         0         1 patient\n",
       "0     1002_CL1  0.961869  0.038131    1002\n",
       "1     1002_CL6  0.995255  0.004745    1002\n",
       "2     1002_CL5  0.992890  0.007110    1002\n",
       "3     1002_CL3  0.987393  0.012607    1002\n",
       "4    1002_CL21  0.203368  0.796632    1002\n",
       "..         ...       ...       ...     ...\n",
       "154   2974_CL3  0.955739  0.044261    2974\n",
       "155   2974_CL4  0.913348  0.086652    2974\n",
       "156   2974_CL5  0.833877  0.166123    2974\n",
       "157   2974_CL6  0.955739  0.044261    2974\n",
       "158   2974_CL9  0.509692  0.490308    2974\n",
       "\n",
       "[159 rows x 4 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmt2c_Pm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "61fc8bd2-e538-4cb1-96fb-bf0f727747e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1326_CL2', '1326_CL9', '1547_CL14', '1547_CL15', '1558_CL1',\n",
       "       '1558_CL10', '2389_CL12', '2389_CL17', '2389_CL18', '2389_CL4',\n",
       "       '2389_CL5', '2389_CL7', '2542_CL12', '2542_CL15', '2542_CL18',\n",
       "       '2542_CL2'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(KMT2C_all_muts['Tumor_Sample_Barcode'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1585c821-5a9c-4aa5-81a7-dfe7b35003c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pm_clones, pm_c_ui, pm_c_idx = np.unique(kmt2c_Pm[\"clone\"], return_inverse = True, return_index = True)\n",
    "\n",
    "sorted_pats, pm_pat_idx, total_clone_per_pat = np.unique(kmt2c_Pm[\"patient\"], return_inverse = True, return_counts=True)\n",
    "pat_to_clone_idx = pm_pat_idx[pm_c_ui]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8b9751dc-1427-45fc-b464-442ea396c27c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1002', '1035', '1045', '1078', '1113', '1326', '1534', '1547',\n",
       "       '1558', '1598', '1644', '2389', '2542', '2819', '2974'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_pats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9c3a5552-064f-4c04-80ff-f8d9e4f8e06f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1326, 1547, 1558, 2389, 2542])"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_patients = np.unique(KMT2C_all_muts['Patient_ID'])\n",
    "mut_patients"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "635733d6-53a2-485d-ad77-58308efbe006",
   "metadata": {},
   "source": [
    "## permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "115836c9-0b97-404a-9b98-74479677bf03",
   "metadata": {},
   "outputs": [],
   "source": [
    "## so I am using the curveball to continue to permute the number of mutations per clone given the number of mutations and clones in the data\n",
    "\n",
    "## OK so my distribution is the probability matrix that I get from the observed\n",
    "## then through the curveball, I permute # of clones mutated and calculate the score\n",
    "## then compare the real score to the observed score under the perms\n",
    "\n",
    "## in this case it is also important to know which clones I am permuting\n",
    "\n",
    "## so i have the whole distribution and vectorized real scores\n",
    "## need to compute scores in a better way"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "cb399520-1c39-4197-8752-dea7c4321350",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "S = subclones_sig_genes\n",
    "\"\"\"\n",
    "_, S[\"qCV\"], _, _ = multitest.multipletests(S[\"pCV\"], method = \"fdr_bh\")\n",
    "subset_idx = (S[\"p\"] < 0.001)\n",
    "\"\"\"\n",
    "\n",
    "subset_idx=(subclones_sig_genes['gene'] =='KMT2C')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "cb933b2e-a61c-4e4c-b096-173eb97cee26",
   "metadata": {},
   "outputs": [],
   "source": [
    "clone_order= list(np.sort(coding_maf[\"Tumor_Sample_Barcode\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0928bfb9-d4ed-4c33-a1fe-aadd682794d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "order_map = {val:i for i,val in enumerate(clone_order)}\n",
    "kmt2c_Pm['clones_sort']=kmt2c_Pm ['clone'].map(order_map)\n",
    "kmt2c_Pm = kmt2c_Pm.sort_values(by='clones_sort')\n",
    "kmt2c_Pm = kmt2c_Pm.drop('clones_sort',axis=1)\n",
    "kmt2c_Pm.reset_index(inplace=True)\n",
    "kmt2c_Pm = kmt2c_Pm.drop('index',axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "ab71b3ef-5204-44e5-9d7e-dd97f50be2c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>patient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1002_CL1</td>\n",
       "      <td>0.961869</td>\n",
       "      <td>0.038131</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1002_CL10</td>\n",
       "      <td>0.872157</td>\n",
       "      <td>0.127843</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1002_CL11</td>\n",
       "      <td>0.991317</td>\n",
       "      <td>0.008683</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1002_CL12</td>\n",
       "      <td>0.932513</td>\n",
       "      <td>0.067487</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002_CL13</td>\n",
       "      <td>0.971077</td>\n",
       "      <td>0.028923</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>2974_CL4</td>\n",
       "      <td>0.913348</td>\n",
       "      <td>0.086652</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>2974_CL5</td>\n",
       "      <td>0.833877</td>\n",
       "      <td>0.166123</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>2974_CL6</td>\n",
       "      <td>0.955739</td>\n",
       "      <td>0.044261</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>2974_CL8</td>\n",
       "      <td>0.396736</td>\n",
       "      <td>0.603264</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>2974_CL9</td>\n",
       "      <td>0.509692</td>\n",
       "      <td>0.490308</td>\n",
       "      <td>2974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>159 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         clone         0         1 patient\n",
       "0     1002_CL1  0.961869  0.038131    1002\n",
       "1    1002_CL10  0.872157  0.127843    1002\n",
       "2    1002_CL11  0.991317  0.008683    1002\n",
       "3    1002_CL12  0.932513  0.067487    1002\n",
       "4    1002_CL13  0.971077  0.028923    1002\n",
       "..         ...       ...       ...     ...\n",
       "154   2974_CL4  0.913348  0.086652    2974\n",
       "155   2974_CL5  0.833877  0.166123    2974\n",
       "156   2974_CL6  0.955739  0.044261    2974\n",
       "157   2974_CL8  0.396736  0.603264    2974\n",
       "158   2974_CL9  0.509692  0.490308    2974\n",
       "\n",
       "[159 rows x 4 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmt2c_Pm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "aaf903b6-ce30-4359-a760-2b0150cbce02",
   "metadata": {},
   "outputs": [],
   "source": [
    "maf = coding_maf\n",
    "## mapping clones to patients\n",
    "clones, c_ui, c_idx = np.unique(maf[\"Tumor_Sample_Barcode\"], return_inverse = True, return_index = True)\n",
    "\n",
    "_, p_idx = np.unique(maf[\"Patient_ID\"], return_inverse = True)\n",
    "c2pat = p_idx[c_ui]\n",
    "\n",
    "# genes\n",
    "genes, g_ui, g_idx = np.unique(maf[\"Hugo_Symbol\"], return_inverse = True, return_index = True)\n",
    "g2gidx = dict(zip(genes, g_idx[g_ui]))\n",
    "\n",
    "#subset_idx_mat = [g2gidx[x] for x in S.loc[subset_idx, \"gene\"]]\n",
    "subset_idx_mat = [g2gidx[x] for x in S.loc[subset_idx, \"gene\"]]\n",
    "\n",
    "# CG is matrix size # total of genes x total # of clones\n",
    "## basically a binarized matrix of the coding maf\n",
    "\n",
    "# c_idx is the indices of the first occurences of the unique values in the original array\n",
    "CG = np.array((sp.coo_matrix((np.ones_like(c_idx), (c_idx, g_idx))).todense() > 0))\n",
    "num_total_mutations = CG.sum(axis=0)\n",
    "n_genes = len(subset_idx_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9b41c615-d073-461d-bb41-d84708570862",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "159"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(c_ui)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "90dc7b1a-f782-4c5e-a07b-f781e1cd4644",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11104"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(g_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "32793249-fa6c-490a-a706-911d9dac275e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11104"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(c_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4c7d9bc8-3054-429e-a28f-e9fe6af38a56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([16])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_num_mutations_in_gene = num_total_mutations[subset_idx_mat]\n",
    "total_num_mutations_in_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4e9cf211-3ae5-486a-9cb6-f963e7a96e3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  0,   0,   0, ..., 158, 158, 158])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "030a55f3-0041-4821-b09c-d3a06a1c2b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[False, False],\n",
       "       [False, False],\n",
       "       [False, False],\n",
       "       [False, False],\n",
       "       [False, False],\n",
       "       [ True, False],\n",
       "       [False, False],\n",
       "       [ True, False],\n",
       "       [ True, False],\n",
       "       [False, False],\n",
       "       [False, False],\n",
       "       [ True, False],\n",
       "       [ True, False]])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def compute_mut_clone_indices(pat_idx, mat):\n",
    "    clone_ids, gene_ids = np.where(mat > 0)  # Get nonzero clone-gene pairs\n",
    "    return npg.aggregate(pat_idx[clone_ids], list(zip(clone_ids, gene_ids)), func='all', axis=0)\n",
    "\n",
    "compute_mut_clone_indices(c2pat, CG[:, subset_idx_mat])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "868a6d2f-ee88-478f-a07b-fb57d2a636aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[54, 64, 73, 79, 89, 90, 106, 107, 109, 110, 112, 117, 122, 130, 133, 136]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## for KMT2C specifically\n",
    "mut_clones = np.unique(KMT2C_all_muts['Tumor_Sample_Barcode'])\n",
    "\n",
    "# map clones back to prob dist\n",
    "mut_clones_in_pm = kmt2c_Pm[kmt2c_Pm[\"clone\"].isin(mut_clones)]\n",
    "mut_clones_in_pm_idx = mut_clones_in_pm.index.tolist()\n",
    "mut_clones_in_pm_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "043541b9-79dd-49dd-a8e8-3ad98a76314e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_mut_clone_indices(pat_idx, mat):\n",
    "    mut_clone_indices = {pat: [] for pat in np.unique(pat_idx)}\n",
    "\n",
    "    for clone_idx, pat in enumerate(pat_idx):\n",
    "        mutated_clones = np.where(mat[clone_idx, :] > 0)[0]  # Get indices of nonzero mutations\n",
    "\n",
    "        mut_clone_indices[pat].extend([clone_idx for gene in mutated_clones])  \n",
    "\n",
    "    return mut_clone_indices\n",
    "\n",
    "mut_clone_indices_obs = compute_mut_clone_indices(c2pat, CG[:, subset_idx_mat])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "039a2db2-df17-462e-9a83-342ee8745e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: [],\n",
       " 1: [],\n",
       " 2: [],\n",
       " 3: [],\n",
       " 4: [],\n",
       " 5: [58, 64],\n",
       " 6: [],\n",
       " 7: [73, 74],\n",
       " 8: [80, 81],\n",
       " 9: [],\n",
       " 10: [],\n",
       " 11: [109, 114, 115, 117, 118, 120],\n",
       " 12: [125, 128, 131, 133],\n",
       " 13: [],\n",
       " 14: []}"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_clone_indices_obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "75cf21f2-deef-4ce6-96f6-130ff70e0572",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_pats = list(kmt2c_Pm['patient'].unique())\n",
    "pat_dict = {val: idx for idx, val in enumerate(unique_pats )}\n",
    "mapped_values = [pat_dict[val] for val in unique_pats ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "12757991-036e-4d67-aec7-8c222b2f5f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_scores(mutated_clone_per_pat):\n",
    "    \"\"\"\n",
    "    input: \n",
    "    gene_pm: probability distribution of a clone having or not having a mutation in given gene, pre-calculated\n",
    "    gene_muts_per_pat: vector of len(total patients) with a 0 at pat index if no gene mutated in that patient, 1 if gene mutated in patient\n",
    "    mut_clones: vector of len(total clones) with a 0 at clone index if no gene mutated in that clone, 1 if gene mutated in clone\n",
    "    pat_indices: array of len(total patients) that represents # of clones in each patient\n",
    "    \n",
    "    returns:\n",
    "    score_vector: vector of len(total patients) with the scores per patient, computed as a sum over the -log of all clones in the patient,\n",
    "    -log( 0<=x<=1 ) + -log(x >= 2 ) per patient\n",
    "    \"\"\"\n",
    "    score_vector = []\n",
    "\n",
    "    for pat in unique_pats:\n",
    "        pat_probs = kmt2c_Pm[kmt2c_Pm['patient']==pat]\n",
    "        running_score=[]\n",
    "        pat_num = pat_dict[pat]\n",
    "        count_muts = len(mutated_clone_per_pat[pat_num])\n",
    "        ## we need 2 or more clones to have the mtuation in order to compute the 1 column\n",
    "\n",
    "        # if 0 <= x <= 1\n",
    "        if count_muts <= 1: \n",
    "            pat_probs_no_KMT2C = pat_probs[0].to_list()\n",
    "            logs_no_KMT2C = [ - np.log(p) for p in pat_probs_no_KMT2C]\n",
    "            running_score.extend(logs_no_KMT2C)\n",
    "\n",
    "        # if x >= 2\n",
    "        if count_muts >=2:\n",
    "            #get indicies of clone\n",
    "            mut_clones_idx = mutated_clone_per_pat[pat_num]\n",
    "\n",
    "            all_clones_in_pat_idx = pat_probs.index.tolist()\n",
    "\n",
    "            for clone_idx in all_clones_in_pat_idx:\n",
    "\n",
    "                if clone_idx in mut_clones_idx:\n",
    "                    pat_probs_clone_KMT2C = []\n",
    "                    pat_probs_test = pat_probs.loc[clone_idx][1]\n",
    "                    pat_probs_clone_KMT2C.append(pat_probs_test)\n",
    "                    logs_KMT2C = [ - np.log(p) for p in pat_probs_clone_KMT2C]\n",
    "                    running_score.extend(logs_KMT2C)\n",
    "                \n",
    "                else:\n",
    "                    ## remove KMT2C clone ids\n",
    "\n",
    "                    pat_probs_no_KMT2C = []\n",
    "\n",
    "                    pat_probs_test = pat_probs.loc[clone_idx][0]\n",
    "                    pat_probs_no_KMT2C.append(pat_probs_test)\n",
    "                    logs_no_KMT2C = [ - np.log(p) for p in pat_probs_no_KMT2C]\n",
    "                    running_score.extend(logs_no_KMT2C)\n",
    "                \n",
    "        score = np.sum(running_score)\n",
    "        score_vector.append(score)\n",
    "    \n",
    "\n",
    "    return np.array(score_vector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "7976d678-f153-445a-87cb-fb6d3e161a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "## my indicies are wrong :( I need to sort the clones the same way in the coding maf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "5c7cb9df-6feb-45c6-84b0-bda523c7d965",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>patient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>1326_CL1</td>\n",
       "      <td>0.779492</td>\n",
       "      <td>0.220508</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>1326_CL10</td>\n",
       "      <td>0.905479</td>\n",
       "      <td>0.094521</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>1326_CL12</td>\n",
       "      <td>0.670312</td>\n",
       "      <td>0.329688</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>1326_CL13</td>\n",
       "      <td>0.362515</td>\n",
       "      <td>0.637485</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>1326_CL2</td>\n",
       "      <td>0.741343</td>\n",
       "      <td>0.258657</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>1326_CL3</td>\n",
       "      <td>0.975526</td>\n",
       "      <td>0.024474</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>1326_CL5</td>\n",
       "      <td>0.975526</td>\n",
       "      <td>0.024474</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>1326_CL6</td>\n",
       "      <td>0.840239</td>\n",
       "      <td>0.159761</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>1326_CL7</td>\n",
       "      <td>0.928278</td>\n",
       "      <td>0.071722</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>1326_CL8</td>\n",
       "      <td>0.975526</td>\n",
       "      <td>0.024474</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>1326_CL9</td>\n",
       "      <td>0.670312</td>\n",
       "      <td>0.329688</td>\n",
       "      <td>1326</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        clone         0         1 patient\n",
       "54   1326_CL1  0.779492  0.220508    1326\n",
       "55  1326_CL10  0.905479  0.094521    1326\n",
       "56  1326_CL12  0.670312  0.329688    1326\n",
       "57  1326_CL13  0.362515  0.637485    1326\n",
       "58   1326_CL2  0.741343  0.258657    1326\n",
       "59   1326_CL3  0.975526  0.024474    1326\n",
       "60   1326_CL5  0.975526  0.024474    1326\n",
       "61   1326_CL6  0.840239  0.159761    1326\n",
       "62   1326_CL7  0.928278  0.071722    1326\n",
       "63   1326_CL8  0.975526  0.024474    1326\n",
       "64   1326_CL9  0.670312  0.329688    1326"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmt2c_Pm[kmt2c_Pm['patient']=='1326']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "9ec191e1-8790-4b2e-8ba7-7cbc4bd35745",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "obs_scores =calculate_scores(mut_clone_indices_obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "938a73dd-33a7-41ca-8631-e4bb1248574a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: [],\n",
       " 1: [],\n",
       " 2: [],\n",
       " 3: [],\n",
       " 4: [],\n",
       " 5: [58, 64],\n",
       " 6: [],\n",
       " 7: [73, 74],\n",
       " 8: [80, 81],\n",
       " 9: [],\n",
       " 10: [],\n",
       " 11: [109, 114, 115, 117, 118, 120],\n",
       " 12: [125, 128, 131, 133],\n",
       " 13: [],\n",
       " 14: []}"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_clone_indices_obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "adf7b222-6c7f-4b83-8f3f-48fd40af8977",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 2.56378182,  2.74361215,  2.78723967,  2.83592867,  2.74375897,\n",
       "        4.54779441,  2.89196222,  2.80316138,  3.40383483,  2.78576736,\n",
       "        2.76256343, 14.71403037,  8.82017986,  2.76006954,  2.67642353])"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "93fd68ae-6df8-4481-938f-5a4f4805544b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(159, 6790)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CG.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f7e1a79d-6d70-4617-8271-a8b908b7a224",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(159, 1)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CG[:, subset_idx_mat].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "85ce42de-220c-463b-97b2-00e218bcb0c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_pat_sums(pat_idx, mat):\n",
    "    return (npg.aggregate(pat_idx, mat, axis = 0) > 0).sum(0)\n",
    "# observed not null, really\n",
    "null_pat_sum = compute_pat_sums(c2pat, CG[:, subset_idx_mat])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "0119f679-4ed0-491c-be66-4bea76c0bd84",
   "metadata": {},
   "outputs": [],
   "source": [
    "## this gets me a list of patients with a 0 or 1 depending on whether a given patient was mutated more than once in the target gene\n",
    "\n",
    "def compute_clone_sums(pat_idx, mat):\n",
    "    return (npg.aggregate(pat_idx, mat, axis = 0) > 1).sum(1)\n",
    "\n",
    "clone_per_patient = compute_clone_sums(c2pat, CG[:, subset_idx_mat])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "fffbb113-7a8b-484d-8cee-32aa1f10dd4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0])"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clone_per_patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "d3035b3e-6fd7-4db3-a32f-1651a8fe25d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0],\n",
       "       [0],\n",
       "       [0],\n",
       "       [0],\n",
       "       [0],\n",
       "       [2],\n",
       "       [0],\n",
       "       [2],\n",
       "       [2],\n",
       "       [0],\n",
       "       [0],\n",
       "       [6],\n",
       "       [4],\n",
       "       [0],\n",
       "       [0]], dtype=int16)"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(npg.aggregate(c2pat, CG[:, subset_idx_mat], axis = 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "4b4c9ecd-56ff-4b4a-959f-a1e92d80e4ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0],\n",
       "       [0],\n",
       "       [0],\n",
       "       [0],\n",
       "       [0],\n",
       "       [2],\n",
       "       [0],\n",
       "       [2],\n",
       "       [2],\n",
       "       [0],\n",
       "       [0],\n",
       "       [6],\n",
       "       [4],\n",
       "       [0],\n",
       "       [0]], dtype=int16)"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## this gets # of clones per patient that are mutated\n",
    "def compute_num_mut_clones(pat_idx, mat):\n",
    "    return (npg.aggregate(c2pat, CG[:, subset_idx_mat], axis = 0))\n",
    "    \n",
    "compute_num_mut_clones(c2pat, CG[:, subset_idx_mat])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1fa1daa5-047b-4a55-a96d-3927c9e27658",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2 2 1 ... 1 3 1]\n"
     ]
    }
   ],
   "source": [
    "gene_patient_matrix = np.zeros_like(CG)\n",
    "for i, pat in enumerate(c2pat):\n",
    "    gene_patient_matrix[pat, :] |= CG[i, :] \n",
    "n_mut_pats = gene_patient_matrix.sum(axis=0)\n",
    "print(n_mut_pats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "d5ad59fc-c380-4b9b-8ea6-283baf2f33e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2.5637818168639304,\n",
       " 2.7436121458661056,\n",
       " 2.787239669333773,\n",
       " 2.8359286696937476,\n",
       " 2.7437589710934858,\n",
       " 4.547794414031646,\n",
       " 2.891962217736308,\n",
       " 2.803161382745473,\n",
       " 3.4038348338438515,\n",
       " 2.7857673611956955,\n",
       " 2.7625634283606044,\n",
       " 14.714030365894862,\n",
       " 8.820179860729802,\n",
       " 2.7600695394391015,\n",
       " 2.6764235305563515]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "545b7f93-ff4f-4218-8460-9316718f20f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.zeros_like( obs_scores))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "ce045224-bb66-4296-8be5-f7dcd5d0749c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15,)"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " obs_scores.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "c31abc47-8d12-460f-8f7e-81f849217293",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15, 1)\n",
      "Chunk 1/10 finished.\n",
      "Chunk 2/10 finished.\n",
      "Chunk 3/10 finished.\n",
      "Chunk 4/10 finished.\n",
      "Chunk 5/10 finished.\n",
      "Chunk 6/10 finished.\n",
      "Chunk 7/10 finished.\n",
      "Chunk 8/10 finished.\n",
      "Chunk 9/10 finished.\n",
      "Chunk 10/10 finished.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\nCb = pd.DataFrame({ \"gene\" : genes[subset_idx_mat], \"p\" : test/(n_chunk*(n_perm//n_chunk)),\\n                      \"n_mut_clones\": num_total_mutations[subset_idx_mat], \\n                    \"n_mut_pats\": n_mut_pats[subset_idx_mat] })\\n'"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_patients = list(coding_maf['Patient_ID'].unique())\n",
    "# precompute nonzero elements\n",
    "hp = curveball.find_presences(CG)\n",
    "\n",
    "#\n",
    "# run permutations (in parallel)\n",
    "pool = multiprocessing.Pool()\n",
    "\n",
    "n_perm = 10000\n",
    "n_chunk = 1000\n",
    "\n",
    "n_genes=1\n",
    "n_iterations = n_perm // n_chunk \n",
    "total_clones = np.arange(1, len(all_patients) )\n",
    "mutation_counts = np.zeros((n_genes, len(all_patients)))\n",
    "test_matrix = np.zeros((n_iterations, len(all_patients)))\n",
    "\n",
    "test = np.zeros((len(obs_scores), 1))\n",
    "print(test.shape)\n",
    "for p_chunk in range(n_perm//n_chunk):\n",
    "    permute_futures = [None]*n_chunk\n",
    "    for p in range(n_chunk):\n",
    "        \n",
    "        permute_futures[p] = pool.apply_async(curveball.curve_ball, (CG, hp, 1000) )\n",
    "    # compute permutation p-values\n",
    "    for p in range(n_chunk):\n",
    "\n",
    "        # one iteration will return a shuffled distribution of clones in patients\n",
    "       \n",
    "        itr_test = ( compute_mut_clone_indices(c2pat, permute_futures[p].get()[:, subset_idx_mat]))\n",
    "\n",
    "        ## then I compute the score of for the given iteration and take the test\n",
    "        scores_per_itr = calculate_scores(itr_test)\n",
    "        \n",
    "        # test should be have a row for every patient\n",
    "        #test += (  calculate_scores(compute_mut_clone_indices(c2pat, permute_futures[p].get()[:, subset_idx_mat])) <= obs_scores )\n",
    "        test += (scores_per_itr >= obs_scores).reshape(len(obs_scores), 1)\n",
    "    \n",
    "    # tracking the final deviation from the observed distribution of # patients with a given mutation on each step         \n",
    "    test_matrix[p_chunk, :] = scores_per_itr.copy()\n",
    "    \n",
    "    print(f\"Chunk {p_chunk + 1}/{n_perm//n_chunk} finished.\")\n",
    "\n",
    "\n",
    "p_values = test / (n_chunk*(n_perm//n_chunk))\n",
    "#Cb = pd.DataFrame({ \"gene\" : genes[subset_idx_mat], \"p\" : test/(n_chunk*(n_perm//n_chunk)) })\n",
    "\n",
    "\"\"\"\n",
    "Cb = pd.DataFrame({ \"gene\" : genes[subset_idx_mat], \"p\" : test/(n_chunk*(n_perm//n_chunk)),\n",
    "                      \"n_mut_clones\": num_total_mutations[subset_idx_mat], \n",
    "                    \"n_mut_pats\": n_mut_pats[subset_idx_mat] })\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70888dfd-1715-4048-9b49-e6a366182967",
   "metadata": {},
   "outputs": [],
   "source": [
    "## I want a P value for every patient actually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "e22359f2-f134-422e-8ff4-268334f2fcf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rx/pxwbtcqd6k38h9yc5d3m4k1w0000gp/T/ipykernel_20715/2260590307.py:1: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)\n",
      "  pd.DataFrame({'patient':list(unique_pats), 'CB_p_vals_KMT2C':list(float(i) for i in p_values)})\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>CB_p_vals_KMT2C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1002</td>\n",
       "      <td>0.9650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1035</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1045</td>\n",
       "      <td>0.9854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1078</td>\n",
       "      <td>0.9923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1113</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1326</td>\n",
       "      <td>0.0146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1534</td>\n",
       "      <td>0.9956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1547</td>\n",
       "      <td>0.2132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1558</td>\n",
       "      <td>0.0704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1598</td>\n",
       "      <td>0.9938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1644</td>\n",
       "      <td>0.9663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2389</td>\n",
       "      <td>0.0122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2542</td>\n",
       "      <td>0.3852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2819</td>\n",
       "      <td>0.9703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2974</td>\n",
       "      <td>0.9881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   patient  CB_p_vals_KMT2C\n",
       "0     1002           0.9650\n",
       "1     1035           1.0000\n",
       "2     1045           0.9854\n",
       "3     1078           0.9923\n",
       "4     1113           1.0000\n",
       "5     1326           0.0146\n",
       "6     1534           0.9956\n",
       "7     1547           0.2132\n",
       "8     1558           0.0704\n",
       "9     1598           0.9938\n",
       "10    1644           0.9663\n",
       "11    2389           0.0122\n",
       "12    2542           0.3852\n",
       "13    2819           0.9703\n",
       "14    2974           0.9881"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({'patient':list(unique_pats), 'CB_p_vals_KMT2C':list(float(i) for i in p_values)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "7bcaac44-38ba-4012-bd1c-5c20dfa90c94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.965 ],\n",
       "       [1.    ],\n",
       "       [0.9854],\n",
       "       [0.9923],\n",
       "       [1.    ],\n",
       "       [0.0146],\n",
       "       [0.9956],\n",
       "       [0.2132],\n",
       "       [0.0704],\n",
       "       [0.9938],\n",
       "       [0.9663],\n",
       "       [0.0122],\n",
       "       [0.3852],\n",
       "       [0.9703],\n",
       "       [0.9881]])"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "5c44c8ab-1f37-422d-9791-a79011a9c5db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "0b2244b0-19f4-4cee-b583-721f5589864f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: [3, 5, 7, 9, 10, 12],\n",
       " 1: [],\n",
       " 2: [],\n",
       " 3: [],\n",
       " 4: [],\n",
       " 5: [64],\n",
       " 6: [],\n",
       " 7: [69, 70],\n",
       " 8: [],\n",
       " 9: [92, 97],\n",
       " 10: [],\n",
       " 11: [109, 112],\n",
       " 12: [125, 129, 130],\n",
       " 13: [],\n",
       " 14: []}"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "itr_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "4ecf99f2-7b79-4027-98be-c6fc9a7674fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 8.2451863 , 10.58791566,  4.47525328,  2.56378182,  8.48429986,\n",
       "       11.39357026,  9.05015   ,  2.56378182,  3.82436107, 15.90810085])"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_matrix.T[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "5ff3bb83-db14-4fe8-ba45-7973ae233890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 5.22393907,  3.95337934,  6.88312278, 10.81665645,  4.88162259,\n",
       "       14.0131802 ,  8.00305622,  4.51553664,  7.97158779,  8.20599173])"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transpose_test[11]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "85e16247-9efb-4ff1-8d0b-5ff9b2156e09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.5637818168639304"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_scores[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ebe02ee-209e-4def-a009-21653d55c15f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
